{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Working jupyter to create function to create a list of aminoacids likely to be used for mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/pdbs\\AF-C0H3V2-F1.pdb\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "path_pdb = './data/pdbs'\n",
    "path_pqr=  './data/pqrs'\n",
    "pdb_file = 'AF-C0H3V2-F1.pdb'\n",
    "output_path = './data/test'\n",
    "\n",
    "print(os.path.join(path_pdb, pdb_file))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fasparse(faspath):\n",
    "    import numpy as np\n",
    "    helixl = []\n",
    "    sheetl = []\n",
    "    data = np.loadtxt(faspath, dtype={'names': ('index', 'col1', 'col2', 'val1', 'val2', 'val3'),'formats': ('i4', 'S1', 'S1', 'f4', 'f4', 'f4')})\n",
    "    indiceshelix = data['index'][data['col2'] == b'H']\n",
    "    indicessheet = data['index'][data['col2'] == b'E']\n",
    "    helix = np.split(indiceshelix, np.where(np.diff(indiceshelix) != 1)[0]+1)\n",
    "    sheet = np.split(indicessheet, np.where(np.diff(indicessheet) != 1)[0]+1)\n",
    "    return [helix, sheet]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "def AA2s4pred (directory_S4pred, output_path, AA_seq, prot):\n",
    "    import os\n",
    "    os.getcwd()\n",
    "    # call s4pred and create fas file\n",
    "    fastapath = os.path.join(output_path,f'{prot}.fasta')\n",
    "    faspath = os.path.join(output_path,f'{prot}.fas')\n",
    "    abs_fasta = os.path.abspath(fastapath)\n",
    "    abs_fas = os.path.abspath(faspath)\n",
    "    print(abs_fasta)\n",
    "    print(abs_fas)    \n",
    "    if os.path.isfile(fastapath):\n",
    "        print(f'fasta file already exists {fastapath}')\n",
    "    else:\n",
    "        with open(os.path.join(output_path,f'{prot}.fasta'), \"w\") as fasta_file:\n",
    "            fasta_file.write(f\">{prot}\\n{AA_seq}\\n\")\n",
    "        print(f'fasta file created {fastapath}')\n",
    "        \n",
    "    if os.path.isfile(faspath):\n",
    "        print('fas file already exists')\n",
    "    else:\n",
    "        os.chdir(directory_S4pred)\n",
    "        os.system(f'python3 run_model.py \"{abs_fasta}\" > \"{abs_fas}\"')\n",
    "        os.chdir('../../')\n",
    "        print('fas file created')\n",
    "     \n",
    "    #read fas file and   \n",
    "    sec_pred = fasparse(abs_fas)    \n",
    "\n",
    "    return sec_pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "c:\\Users\\marik\\OneDrive - bwedu\\Uni HD\\FS 4\\Bioinfo\\topic04_02\\data\\test\\test_tobi.fasta\n",
      "c:\\Users\\marik\\OneDrive - bwedu\\Uni HD\\FS 4\\Bioinfo\\topic04_02\\data\\test\\test_tobi.fas\n",
      "fasta file already exists ./data/test\\test_tobi.fasta\n",
      "fas file already exists\n",
      "[array([13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27]), array([47, 48, 49, 50, 51, 52, 53, 54, 55]), array([70, 71, 72, 73, 74, 75, 76, 77, 78]), array([124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134]), array([151, 152, 153, 154, 155, 156, 157, 158, 159]), array([204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214])]\n"
     ]
    }
   ],
   "source": [
    "output_path = './data/test'\n",
    "AA_seq = 'MRVLFIGDVMAEPGLRAVGLHLPDIRDRYDLVIANGENAARGKGLDRRSYRLLREAGVDLVSLGNHAWDHKEVYALLESEPVVRPLNYPPGTPGKGFWRLEVGGESLLFVQVMGRIFMDPLDDPFRALDRLLEEEKADYVLVEVHAEATSEKMALAHYLDGRASAVLGTHTHVPTLDATRLPKGTLYQTDVGMTGTYHSIIGGEVETFLARFLTGRPQPFRAAQGKARLHATELVFEGGRPVAISPYVWEEP'\n",
    "test = AA2s4pred('./data/S4pred', output_path, AA_seq, prot='test_tobi') #outerarray 0 is helix, #outer_array 1 is sheet\n",
    "helix = test[0]\n",
    "print(helix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prot_mut(pdb_path, pdb_file, pqr_output_path, Deep_mut=True, iterations=1, cutoff_value = -3):\n",
    "    \n",
    "    import os\n",
    "    from helper_function import pdb2AA\n",
    "    from function import functional_aa\n",
    "    from function import free_aa\n",
    "    import numpy as np\n",
    "    \n",
    "    \n",
    "    # top 20 features that correlate with melt point (positive and negative), based on prokaryotes323 columns, cutoff at +/- 0.32\n",
    "    pos_corr = {'YR': 0.492442, \n",
    "                'RP': 0.480594, \n",
    "                'RG': 0.442533, \n",
    "                'R': 0.432905, \n",
    "                'WR': 0.392523, \n",
    "                'YP': 0.390709, \n",
    "                'LR': 0.386596, \n",
    "                'FR': 0.376553, \n",
    "                'VR': 0.370241, \n",
    "                'ER': 0.369174, \n",
    "                'RC': 0.357052, \n",
    "                #'Rhelix': 0.349204, # percentage of R in all helices\n",
    "                'MR': 0.343496, \n",
    "                'P': 0.340082, \n",
    "                'PG': 0.338434, \n",
    "                'EAmotif': 0.332202, # EA hinterenander\n",
    "                'LP': 0.330337, \n",
    "                'EP': 0.328569, \n",
    "                'RH': 0.326193, \n",
    "                'AR': 0.324355, \n",
    "                'ARmotif': 0.3241, # AR hinterenander\n",
    "                'NR': 0.323378}\n",
    "    \n",
    "    neg_corr = {'QT': -0.52935, \n",
    "                'MQ': -0.507329, \n",
    "                'QS': -0.502697, \n",
    "                'QC': -0.493738, \n",
    "                'Q': -0.469765, \n",
    "                'QD': -0.466556, \n",
    "                'QH': -0.455041, \n",
    "                'NQ': -0.435562, \n",
    "                'IQ': -0.429683, \n",
    "                'FQ': -0.42363, \n",
    "                'WQ': -0.420057, \n",
    "                'QK': -0.41872, \n",
    "                #'PolarAA': -0.406035, \n",
    "                'ST': -0.396185, \n",
    "                #'Qhelix': -0.37895, \n",
    "                'TC': -0.364763, \n",
    "                'MT': -0.359458, \n",
    "                'TH': -0.346921, \n",
    "                'TD': -0.34469, \n",
    "                'SH': -0.327684, \n",
    "                'SC': -0.321686, \n",
    "                'T': -0.321208}\n",
    "    sorted_freq_pos = ['R', 'P', 'L', 'Y', 'E', 'G', 'A', 'V', 'M', 'F', 'W', 'N', 'H', 'C']\n",
    "    sorted_freq_neg = ['K', 'N', 'W', 'F', 'I', 'D', 'M', 'C', 'H', 'S', 'T', 'Q']\n",
    "\n",
    "\n",
    "    ideal_pos_value = {'AR': 0.19577028919092312, 'VR': 0.16570774850840858, 'LR': 0.22057443760531548, 'LP': 0.19913083779456836, 'MR': 0.10189569954318338, 'FR': 0.1214807452476434, 'WR': 0.09732041140609597, 'NR': 0.10350733491352844, 'YR': 0.11505796446768121, 'YP': 0.09361436465693411, 'ER': 0.1750096588731577, 'EP': 0.15356605906241058, 'RH': 0.10505828276892132, 'RC': 0.0895071066024632, 'RP': 0.14867680267495723, 'RG': 0.1743851008086472, 'PG': 0.15294150099790005, 'R': 0.08506020124285214, 'P': 0.06361660143210504, 'ARmotif': 0.010758762261349177, 'EAmotif': 0.01633251571740483}\n",
    "    ideal_neg_value = {'IQ': 0.05490211183163266, 'MQ': 0.03973022298244349, 'MT': 0.054568274035922376, 'FQ': 0.05931526868690351, 'WQ': 0.03515493484535609, 'NQ': 0.041341858352788544, 'QS': 0.057844351285310555, 'QT': 0.06062750041770345, 'QD': 0.06216042713743332, 'QH': 0.04289280620818142, 'QK': 0.06281534035966867, 'QC': 0.027341630041723304, 'ST': 0.07268240233878943, 'SH': 0.054947708129267414, 'SC': 0.03939653196280929, 'TD': 0.0769984781909122, 'TH': 0.0577308572616603, 'TC': 0.04217968109520218, 'Q': 0.022894724682112268, 'T': 0.03773277573559115}\n",
    "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        \n",
    "\n",
    "    #https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003674\n",
    "    conserv_subst = {\n",
    "        'A': ['D', 'E', 'G', 'S', 'T'],\n",
    "        'C': ['G', 'R', 'S', 'W', 'Y'],\n",
    "        'D': ['A', 'E', 'G', 'H', 'N', 'V', 'Y'],\n",
    "        'E': ['A', 'D', 'G', 'K', 'Q', 'V'],\n",
    "        'F': ['I', 'L', 'Y'],\n",
    "        'G': ['A', 'C', 'D', 'E', 'R'],\n",
    "        'H': ['D', 'L', 'N', 'P', 'Q', 'R', 'Y'],\n",
    "        'I': ['F', 'L', 'M', 'N', 'V'],\n",
    "        'K': ['E', 'M', 'N', 'Q', 'R', 'T'],\n",
    "        'L': ['F', 'H', 'I', 'M', 'P', 'Q', 'R', 'V', 'W'],\n",
    "        'M': ['I', 'K', 'L', 'R', 'T', 'V'],\n",
    "        'N': ['D', 'H', 'I', 'K', 'S', 'T', 'Y'],\n",
    "        'P': ['H', 'L', 'Q', 'R', 'S'],\n",
    "        'Q': ['E', 'H', 'K', 'L', 'P', 'R'],\n",
    "        'R': ['C', 'G', 'H', 'K', 'L', 'M', 'P', 'Q', 'T', 'W'],\n",
    "        'S': ['A', 'C', 'N', 'P', 'T', 'W', 'Y'],\n",
    "        'T': ['A', 'K', 'M', 'N', 'R', 'S'],\n",
    "        'V': ['D', 'E', 'I', 'L', 'M'],\n",
    "        'W': ['C', 'L', 'R', 'S'],\n",
    "        'Y': ['C', 'D', 'F', 'H', 'N'],\n",
    "        }\n",
    "\n",
    "    non_conservative_substitutions = {\n",
    "        'A': ['P', 'V'],\n",
    "        'C': ['F'],\n",
    "        'F': ['C', 'S', 'V'],\n",
    "        'G': ['S', 'V', 'W'],\n",
    "        'I': ['K', 'R', 'S', 'T'],\n",
    "        'K': ['I'],\n",
    "        'L': ['S'],\n",
    "        'P': ['A', 'T'],\n",
    "        'Q': ['E', 'K'],\n",
    "        'R': ['I', 'S'],\n",
    "        'S': ['F', 'G', 'I', 'L', 'R'],\n",
    "        'T': ['I', 'P'],\n",
    "        'V': ['A', 'F', 'G'],\n",
    "        'W': ['G'],\n",
    "        }\n",
    "\n",
    "    #extract aminoacid list\n",
    "    aa_list = pdb2AA(pdb_path, pdb_file)\n",
    "    aa_locked = functional_aa(pdb_path, pdb_file, pqr_output_path)\n",
    "    aa_free = free_aa(pdb_path, pdb_file, aa_locked)\n",
    "    AAs = ''.join(aa_list)\n",
    "    # calculate ratio of WT protein for pos_corr and neg_corr and sort them in descending order\n",
    "    from function import rel_aa_comp\n",
    "    from function_mut import diff_weighted\n",
    "    from function_mut import mut_apply\n",
    "    from function_mut import mutator2\n",
    "    from helper_function import ArraySlice\n",
    "    WT_dev_sum, WT_dev = diff_weighted(pos_corr, neg_corr, aa_list, ideal_pos_value, ideal_neg_value)\n",
    "    \n",
    "    #define variables for iteration\n",
    "    prev_Mut_prot_list = aa_list         #list of all aminoacids in the protein\n",
    "    prev_Mut_dev_sum = WT_dev_sum   # Sum of deviation of protein sequence from ideal values\n",
    "    prev_Mut_dev = WT_dev           # dictionary with deviation from ideal from each feature\n",
    "    aa_available = aa_free          # list of aminoacids that can be mutated\n",
    "    \n",
    "    best_Mut_prot_list = aa_list\n",
    "    best_Mut_dev_sum = WT_dev_sum\n",
    "    best_Mut_dev = WT_dev\n",
    "    best_aa_available = aa_available\n",
    "    \n",
    "    Dev_list =[WT_dev]\n",
    "\n",
    "    cutoff = 0\n",
    "    while cutoff >= cutoff_value:\n",
    "        print (f'##################### cutoff: {cutoff} #####################')\n",
    "        #resets value to original values\n",
    "        prev_Mut_prot_list = aa_list         #list of all aminoacids in the protein\n",
    "        prev_Mut_dev_sum = WT_dev_sum   # Sum of deviation of protein sequence from ideal values\n",
    "        prev_Mut_dev = WT_dev           # dictionary with deviation from ideal from each feature\n",
    "        aa_available = aa_free          # list of aminoacids that can be mutated\n",
    "        \n",
    "        for i in range(iterations):\n",
    "            Mut_prot_list, possible_mutations = mutator3(\n",
    "                                                    AA_list = prev_Mut_prot_list, \n",
    "                                                    free_AA = aa_available, \n",
    "                                                    deviation = prev_Mut_dev,\n",
    "                                                    pos_corr = pos_corr, \n",
    "                                                    neg_corr =  neg_corr, \n",
    "                                                    conserv_substitution = conserv_subst,\n",
    "                                                    ideal_pos_value = ideal_pos_value, \n",
    "                                                    ideal_neg_value = ideal_neg_value,\n",
    "                                                    cutoff_value = cutoff,\n",
    "                                                    deep_mut = Deep_mut) # calculates list of possible mutations\n",
    "            \n",
    "            Mut_dev_sum, Mut_dev = diff_weighted(pos_corr, neg_corr, Mut_prot_list, ideal_pos_value, ideal_neg_value) # calculate deviation of mutated protein sequence\n",
    "            Dev_list.append(Mut_dev)\n",
    "            \n",
    "            if abs(best_Mut_dev_sum) > abs(Mut_dev_sum):\n",
    "                best_Mut_prot_list = Mut_prot_list\n",
    "                best_Mut_dev_sum = Mut_dev_sum\n",
    "                best_Mut_dev = Mut_dev\n",
    "                best_aa_available = aa_available\n",
    "                best_possible_mutations = possible_mutations\n",
    "                output_message = f' Current state:\\n cutoff value: {cutoff} \\n Best improvement after {i+1} iterations \\n WT deviation:   {WT_dev_sum} \\n Best deviation: {best_Mut_dev_sum} \\n improved by {best_Mut_dev_sum - WT_dev_sum} \\n current cutoff value = {cutoff}'\n",
    "            elif Mut_dev[0][0] == prev_Mut_dev[0][0]:\n",
    "                break\n",
    "                \n",
    "            \n",
    "            #update variables for next iteration            \n",
    "            prev_possible_mutations = possible_mutations    #list of mutations (AA-POS-AA) (prev_possible_mutations can be printed if needed)\n",
    "            prev_Mut_prot_list = Mut_prot_list                        #Mutated protein as a list with one AA per entry\n",
    "            prev_Mut_dev = Mut_dev\n",
    "            prev_Mut_dev_sum = Mut_dev_sum\n",
    "            aa_available = ArraySlice(aa_available, possible_mutations) #updates available aminoacids, so that each aminoacid can only be mutated once\n",
    "        print(f'!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: {cutoff} !!!!!!!!!!!!!!!!!!!!!!!!!!!!')\n",
    "        cutoff += -0.1\n",
    "            \n",
    "    \n",
    "    print( f'########## \\n {output_message} \\n ##########')  \n",
    "    \n",
    "\n",
    "    Improvement = best_Mut_dev_sum - WT_dev_sum\n",
    "    #return Improvement, Dev_list, prev_Mut_prot_list, AAs\n",
    "    return WT_dev_sum, WT_dev, best_Mut_dev_sum, best_Mut_dev, Improvement, best_possible_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "def random_mutate(AA_list, free_AA, mutation_dict, mutation_rate):\n",
    "    \"\"\"\n",
    "    Mutates a protein sequence randomly, respecting allowed mutations from a dictionary.\n",
    "\n",
    "    Args:\n",
    "        sequence (str): The protein sequence to mutate.\n",
    "        mutation_dict (dict): Dictionary where keys are amino acids and values are lists of possible mutations.\n",
    "\n",
    "    Returns:\n",
    "        str: The mutated protein sequence.\n",
    "    \"\"\"\n",
    "    import random\n",
    "    from function_mut import mut_apply\n",
    "    mut_list = []\n",
    "    aa_mut_list = AA_list\n",
    "    free_AA_dict = {a: b for a, b in zip(free_AA[:,2], free_AA[:,1] )} # create dictionary from array the key is the absolute aminoacid position and the value the aminoacid\n",
    "\n",
    "    for key in free_AA_dict:\n",
    "        aminoacid = key\n",
    "        if aminoacid in mutation_dict:\n",
    "            possible_mutation = mutation_dict[aminoacid]\n",
    "            if random.random() < mutation_rate:\n",
    "                mutation = random.choice(possible_mutation)\n",
    "                mut_aa = f'{aminoacid} - {free_AA_dict[key]} - {mutation}'\n",
    "                mut_list.append(mut_aa)\n",
    "                aa_mut_list = mut_apply(AA_list, [mut_aa])\n",
    "    return aa_mut_list, mut_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pqr file already exists\n",
      "Salt_bridge finished\n",
      "H_bond finished\n",
      "VdW_interaction finished\n",
      "##################### cutoff: 0 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.19\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.18\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.16999999999999998\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.15999999999999998\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.14999999999999997\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.13999999999999996\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.12999999999999995\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.11999999999999995\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.10999999999999996\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.09999999999999996\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.08999999999999997\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.07999999999999997\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\marik\\OneDrive - bwedu\\Uni HD\\FS 4\\Bioinfo\\topic04_02\\helper_function.py:71: RuntimeWarning: invalid value encountered in cast\n",
      "  distance[:,0] = distance[:,0].astype('int')\n",
      "c:\\Users\\marik\\OneDrive - bwedu\\Uni HD\\FS 4\\Bioinfo\\topic04_02\\helper_function.py:89: RuntimeWarning: invalid value encountered in arccos\n",
      "  theta = np.arccos((d_DH[n,1,1]**2 + d_HA[n,1:]**2 - d_DA[1:,1:][n,:]**2)/(2*d_DH[n,1,1]*d_HA[1:,1:][n,:]))\n",
      "c:\\Users\\marik\\OneDrive - bwedu\\Uni HD\\FS 4\\Bioinfo\\topic04_02\\helper_function.py:106: RuntimeWarning: invalid value encountered in cast\n",
      "  array[:,0] = array[:,0].astype('int')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.06999999999999998\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.05999999999999998\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.049999999999999975\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.03999999999999997\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.02999999999999997\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.01999999999999997\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.009999999999999969\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -3.122502256758253e-17\n",
      "current level of deviation: ('IQ', 0.18442966495731225) -- IQ --- cutoff 0.19\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- IQ --- cutoff 0.18\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- IQ --- cutoff 0.16999999999999998\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- IQ --- cutoff 0.15999999999999998\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- IQ --- cutoff 0.14999999999999997\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- IQ --- cutoff 0.13999999999999996\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- IQ --- cutoff 0.12999999999999995\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- IQ --- cutoff 0.11999999999999995\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- IQ --- cutoff 0.10999999999999996\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- IQ --- cutoff 0.09999999999999996\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- IQ --- cutoff 0.08999999999999997\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- IQ --- cutoff 0.07999999999999997\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- IQ --- cutoff 0.06999999999999998\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- IQ --- cutoff 0.05999999999999998\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.19\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.18\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.16999999999999998\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.15999999999999998\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.14999999999999997\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.13999999999999996\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.12999999999999995\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.11999999999999995\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.10999999999999996\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.09999999999999996\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.08999999999999997\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.07999999999999997\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.06999999999999998\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.05999999999999998\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.049999999999999975\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.03999999999999997\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.02999999999999997\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.01999999999999997\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.009999999999999969\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff -3.122502256758253e-17\n",
      "current level of deviation: ('FQ', 0.19295989510600078) -- FQ --- cutoff 0.19\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.18\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.16999999999999998\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.15999999999999998\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.14999999999999997\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.13999999999999996\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.12999999999999995\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.11999999999999995\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.10999999999999996\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.09999999999999996\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.08999999999999997\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.07999999999999997\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.06999999999999998\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.05999999999999998\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.049999999999999975\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.03999999999999997\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.02999999999999997\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.01999999999999997\n",
      "current level of deviation: ('P', 0.0006558400692121651) -- FQ --- cutoff 0.009999999999999969\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.19\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.18\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.16999999999999998\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.15999999999999998\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.14999999999999997\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.13999999999999996\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.12999999999999995\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.11999999999999995\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.10999999999999996\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.09999999999999996\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.08999999999999997\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.07999999999999997\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.06999999999999998\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.05999999999999998\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.049999999999999975\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.03999999999999997\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.02999999999999997\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.01999999999999997\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff 0.009999999999999969\n",
      "current level of deviation: ('EAmotif', 0.24632255736189845) -- EAmotif --- cutoff -3.122502256758253e-17\n",
      "current level of deviation: ('FQ', 0.1579948601409658) -- FQ --- cutoff 0.19\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.18\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.16999999999999998\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.15999999999999998\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.14999999999999997\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.13999999999999996\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.12999999999999995\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.11999999999999995\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.10999999999999996\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.09999999999999996\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.08999999999999997\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.07999999999999997\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.06999999999999998\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.05999999999999998\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.049999999999999975\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.03999999999999997\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.02999999999999997\n",
      "current level of deviation: ('MR', 3.5730245250352455e-05) -- FQ --- cutoff 0.01999999999999997\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: 0 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -0.1 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.09000000000000001\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.08000000000000002\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.07000000000000002\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.06000000000000002\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.05000000000000002\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.040000000000000015\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.030000000000000013\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.02000000000000001\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 0.01000000000000001\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff 1.0408340855860843e-17\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -0.00999999999999999\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -0.1 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -0.2 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -0.01\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -0.2 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -0.30000000000000004 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -0.11000000000000003\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -0.30000000000000004 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -0.4 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -0.21000000000000002\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -0.4 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -0.5 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -0.31\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -0.5 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -0.6 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -0.41\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -0.6 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -0.7 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -0.5099999999999999\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -0.7 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -0.7999999999999999 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -0.6099999999999999\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -0.7999999999999999 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -0.8999999999999999 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -0.71\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -0.8999999999999999 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -0.9999999999999999 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -0.8099999999999998\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -0.9999999999999999 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -1.0999999999999999 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -0.9099999999999999\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -1.0999999999999999 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -1.2 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -1.01\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -1.2 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -1.3 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -1.11\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -1.3 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -1.4000000000000001 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -1.2100000000000002\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -1.4000000000000001 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -1.5000000000000002 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -1.3100000000000003\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -1.5000000000000002 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -1.6000000000000003 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -1.4100000000000004\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -1.6000000000000003 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -1.7000000000000004 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -1.5100000000000005\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -1.7000000000000004 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -1.8000000000000005 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -1.6100000000000005\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -1.8000000000000005 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -1.9000000000000006 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -1.7100000000000006\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -1.9000000000000006 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -2.0000000000000004 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -1.8100000000000005\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -2.0000000000000004 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -2.1000000000000005 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -1.9100000000000006\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -2.1000000000000005 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -2.2000000000000006 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -2.0100000000000002\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -2.2000000000000006 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -2.3000000000000007 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -2.1100000000000003\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -2.3000000000000007 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -2.400000000000001 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -2.2100000000000004\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -2.400000000000001 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -2.500000000000001 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -2.3100000000000005\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -2.500000000000001 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -2.600000000000001 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -2.4100000000000006\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -2.600000000000001 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -2.700000000000001 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -2.5100000000000007\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -2.700000000000001 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -2.800000000000001 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -2.6100000000000008\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -2.800000000000001 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -2.9000000000000012 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -2.710000000000001\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -2.9000000000000012 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -3.0000000000000013 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -2.810000000000001\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -3.0000000000000013 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -3.1000000000000014 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -2.910000000000001\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -3.1000000000000014 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -3.2000000000000015 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -3.010000000000001\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -3.2000000000000015 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -3.3000000000000016 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -3.110000000000001\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -3.3000000000000016 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -3.4000000000000017 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -3.2100000000000013\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -3.4000000000000017 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -3.5000000000000018 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -3.3100000000000014\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -3.5000000000000018 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -3.600000000000002 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -3.4100000000000015\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -3.600000000000002 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -3.700000000000002 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -3.5100000000000016\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -3.700000000000002 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -3.800000000000002 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -3.6100000000000017\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -3.800000000000002 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -3.900000000000002 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -3.7100000000000017\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -3.900000000000002 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -4.000000000000002 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -3.8100000000000014\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -4.000000000000002 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -4.100000000000001 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -3.910000000000001\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -4.100000000000001 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -4.200000000000001 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -4.010000000000001\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -4.200000000000001 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -4.300000000000001 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -4.11\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -4.300000000000001 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -4.4 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -4.21\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -4.4 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -4.5 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -4.31\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -4.5 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -4.6 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -4.409999999999999\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -4.6 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -4.699999999999999 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -4.509999999999999\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -4.699999999999999 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -4.799999999999999 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -4.6099999999999985\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -4.799999999999999 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -4.899999999999999 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -4.709999999999998\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -4.899999999999999 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "##################### cutoff: -4.999999999999998 #####################\n",
      "current level of deviation: ('EAmotif', 0.21135752239686345) -- EAmotif --- cutoff -4.809999999999998\n",
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!! Iteration completed: cutoff value: -4.999999999999998 !!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "########## \n",
      "  Current state:\n",
      " cutoff value: 0 \n",
      " Best improvement after 2 iterations \n",
      " WT deviation:   2.698185915449055 \n",
      " Best deviation: 2.5458647421835376 \n",
      " improved by -0.1523211732655172 \n",
      " current cutoff value = 0 \n",
      " ##########\n",
      "2.698185915449055\n",
      "[('EAmotif', 0.21135752239686345), ('IQ', 0.18442966495731225), ('ST', 0.16166281043772615), ('QS', 0.14096629374618364), ('NQ', 0.12290204740296218), ('SH', 0.12290058968573575), ('QK', 0.11721113022449137), ('EP', 0.11038211437908366), ('QT', 0.10901624426918824), ('SC', 0.10358240369007918), ('TD', 0.10346368237070246), ('ER', 0.09623014503432172), ('TH', 0.08995811666214239), ('MT', 0.08954507257847452), ('FQ', 0.08806479021089587), ('QD', 0.08494537618758828), ('ARmotif', 0.08042916906718064), ('TC', 0.07132964295938518), ('MQ', 0.06910734324650902), ('QH', 0.06846903438082604), ('T', 0.06107111837952671), ('NR', 0.05743709595942391), ('QC', 0.04846463669857535), ('RP', 0.04348115133274043), ('WQ', 0.04273910443836371), ('Q', 0.03872716831232044), ('PG', 0.032155479967360616), ('WR', 0.02421448586034103), ('R', 0.02283697243302292), ('YR', 0.021694339173358904), ('AR', 0.020417011765562043), ('LP', 0.011142793358532604), ('RC', 0.010979670447601706), ('VR', 0.008578267414568234), ('LR', 0.00835011835738761), ('P', 0.007648847062219159), ('RG', 0.006744922099930856), ('FR', 0.0032071098858130823), ('YP', 0.0016109398357110963), ('RH', 0.0006957585337922165), ('MR', 3.5730245250352455e-05)]\n",
      "2.5458647421835376\n"
     ]
    }
   ],
   "source": [
    "test_list = ['AF-P0AGD1-F1.pdb', 'AF-O34633-F1.pdb', 'AF-Q72L06-F1.pdb', 'AF-Q746J6-F1.pdb', 'AF-R4YU54-F1.pdb', 'AF-Q72L88-F1.pdb','AF-P21340-F1.pdb' , 'AF-Q745V2-F1.pdb', 'AF-C0H3V2-F1.pdb']\n",
    "pdb_file = test_list[8]\n",
    "Dev_sum, Dev, mut_dev, mut_sum, Improv, possible_mut = prot_mut(path_pdb, pdb_file, path_pqr, Deep_mut=True, iterations=4, cutoff_value = -5)\n",
    "print(Dev_sum)\n",
    "print(Dev)\n",
    "print(mut_dev)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['F-44-L']\n"
     ]
    }
   ],
   "source": [
    "print(possible_mut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pqr file already exists\n",
      "Salt_bridge finished\n",
      "H_bond finished\n",
      "VdW_interaction finished\n",
      "##################### cutoff: 0.1 #####################\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.29000000000000004\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.28\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.27\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.26\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.25\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.24\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.22999999999999998\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.21999999999999997\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.20999999999999996\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.19999999999999996\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.18999999999999995\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.17999999999999994\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.16999999999999993\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.15999999999999992\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\marik\\OneDrive - bwedu\\Uni HD\\FS 4\\Bioinfo\\topic04_02\\helper_function.py:71: RuntimeWarning: invalid value encountered in cast\n",
      "  distance[:,0] = distance[:,0].astype('int')\n",
      "c:\\Users\\marik\\OneDrive - bwedu\\Uni HD\\FS 4\\Bioinfo\\topic04_02\\helper_function.py:106: RuntimeWarning: invalid value encountered in cast\n",
      "  array[:,0] = array[:,0].astype('int')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.1499999999999999\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.1399999999999999\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.1299999999999999\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.1199999999999999\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.1099999999999999\n",
      "current level of deviation: ('EAmotif', 0.16798471023792413) -- EAmotif --- cutoff 0.09999999999999991\n",
      "current level of deviation: ('PG', 0.14477118795879115) -- PG --- cutoff 0.29000000000000004\n",
      "current level of deviation: ('RC', 0.0030569573225764438) -- PG --- cutoff 0.28\n",
      "current level of deviation: ('RC', 0.0030569573225764438) -- PG --- cutoff 0.27\n",
      "current level of deviation: ('RC', 0.0030569573225764438) -- PG --- cutoff 0.26\n",
      "current level of deviation: ('RC', 0.0030569573225764438) -- PG --- cutoff 0.25\n",
      "current level of deviation: ('RC', 0.0030569573225764438) -- PG --- cutoff 0.24\n",
      "current level of deviation: ('RC', 0.0030569573225764438) -- PG --- cutoff 0.22999999999999998\n",
      "current level of deviation: ('RC', 0.0030569573225764438) -- PG --- cutoff 0.21999999999999997\n",
      "current level of deviation: ('RC', 0.0030569573225764438) -- PG --- cutoff 0.20999999999999996\n",
      "current level of deviation: ('RC', 0.0030569573225764438) -- PG --- cutoff 0.19999999999999996\n",
      "current level of deviation: ('RC', 0.0030569573225764438) -- PG --- cutoff 0.18999999999999995\n",
      "current level of deviation: ('RC', 0.0030569573225764438) -- PG --- cutoff 0.17999999999999994\n",
      "current level of deviation: ('RC', 0.0030569573225764438) -- PG --- cutoff 0.16999999999999993\n",
      "current level of deviation: ('RC', 0.0030569573225764438) -- PG --- cutoff 0.15999999999999992\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[252], line 7\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m n \u001b[38;5;129;01min\u001b[39;00m test_list:\n\u001b[0;32m      6\u001b[0m     pdb_file \u001b[38;5;241m=\u001b[39m n\n\u001b[1;32m----> 7\u001b[0m     Dev_sum, Dev, Mut_sum, Mut_dev, Improvement, possible_mut \u001b[38;5;241m=\u001b[39m \u001b[43mprot_mut\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpath_pdb\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpdb_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpath_pqr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mDeep_mut\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43miterations\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m5\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcutoff_value\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m      8\u001b[0m     results[n] \u001b[38;5;241m=\u001b[39m Improvement\n\u001b[0;32m      9\u001b[0m     Possible_mut[n] \u001b[38;5;241m=\u001b[39m possible_mut\n",
      "Cell \u001b[1;32mIn[249], line 141\u001b[0m, in \u001b[0;36mprot_mut\u001b[1;34m(pdb_path, pdb_file, pqr_output_path, Deep_mut, iterations, cutoff_value)\u001b[0m\n\u001b[0;32m    138\u001b[0m aa_available \u001b[38;5;241m=\u001b[39m aa_free          \u001b[38;5;66;03m# list of aminoacids that can be mutated\u001b[39;00m\n\u001b[0;32m    140\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(iterations):\n\u001b[1;32m--> 141\u001b[0m     Mut_prot_list, possible_mutations \u001b[38;5;241m=\u001b[39m \u001b[43mmutator3\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    142\u001b[0m \u001b[43m                                            \u001b[49m\u001b[43mAA_list\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mprev_Mut_prot_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[0;32m    143\u001b[0m \u001b[43m                                            \u001b[49m\u001b[43mfree_AA\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43maa_available\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[0;32m    144\u001b[0m \u001b[43m                                            \u001b[49m\u001b[43mdeviation\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mprev_Mut_dev\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    145\u001b[0m \u001b[43m                                            \u001b[49m\u001b[43mpos_corr\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mpos_corr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[0;32m    146\u001b[0m \u001b[43m                                            \u001b[49m\u001b[43mneg_corr\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m  \u001b[49m\u001b[43mneg_corr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[0;32m    147\u001b[0m \u001b[43m                                            \u001b[49m\u001b[43mconserv_substitution\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mconserv_subst\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    148\u001b[0m \u001b[43m                                            \u001b[49m\u001b[43mideal_pos_value\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mideal_pos_value\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[0;32m    149\u001b[0m \u001b[43m                                            \u001b[49m\u001b[43mideal_neg_value\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mideal_neg_value\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    150\u001b[0m \u001b[43m                                            \u001b[49m\u001b[43mcutoff_value\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mcutoff\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    151\u001b[0m \u001b[43m                                            \u001b[49m\u001b[43mdeep_mut\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mDeep_mut\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;66;03m# calculates list of possible mutations\u001b[39;00m\n\u001b[0;32m    153\u001b[0m     Mut_dev_sum, Mut_dev \u001b[38;5;241m=\u001b[39m diff_weighted(pos_corr, neg_corr, Mut_prot_list, ideal_pos_value, ideal_neg_value) \u001b[38;5;66;03m# calculate deviation of mutated protein sequence\u001b[39;00m\n\u001b[0;32m    154\u001b[0m     Dev_list\u001b[38;5;241m.\u001b[39mappend(Mut_dev)\n",
      "Cell \u001b[1;32mIn[251], line 113\u001b[0m, in \u001b[0;36mmutator3\u001b[1;34m(AA_list, free_AA, deviation, pos_corr, neg_corr, conserv_substitution, ideal_pos_value, ideal_neg_value, cutoff_value, deep_mut)\u001b[0m\n\u001b[0;32m    111\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(subst) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m: \u001b[38;5;66;03m# if only one substitution increases one of the aminoacids this substitution will be used\u001b[39;00m\n\u001b[0;32m    112\u001b[0m     mut_aa \u001b[38;5;241m=\u001b[39m aminoacid \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m key \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m subst\n\u001b[1;32m--> 113\u001b[0m     Diff \u001b[38;5;241m=\u001b[39m \u001b[43mmut_live_test\u001b[49m\u001b[43m(\u001b[49m\u001b[43mAA_mut_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[43mmut_aa\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpos_corr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mneg_corr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mideal_pos_value\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mideal_neg_value\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;66;03m# live test if mutation is benefitical\u001b[39;00m\n\u001b[0;32m    114\u001b[0m     mut_rel \u001b[38;5;241m=\u001b[39m rel_aa_comp(AAs, subst)\n\u001b[0;32m    115\u001b[0m     mut_rel_max \u001b[38;5;241m=\u001b[39m rel_aa_comp(AAs, subst) \u001b[38;5;241m*\u001b[39m max_increase\n",
      "File \u001b[1;32mc:\\Users\\marik\\OneDrive - bwedu\\Uni HD\\FS 4\\Bioinfo\\topic04_02\\function_mut.py:86\u001b[0m, in \u001b[0;36mmut_live_test\u001b[1;34m(AA_list, Mut_list, pos_corr, neg_corr, ideal_pos_value, ideal_neg_value)\u001b[0m\n\u001b[0;32m     84\u001b[0m AAs \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(AA_list)\n\u001b[0;32m     85\u001b[0m WT_sum, WT_diff \u001b[38;5;241m=\u001b[39m diff_weighted(pos_corr, neg_corr, AAs, ideal_pos_value, ideal_neg_value)\n\u001b[1;32m---> 86\u001b[0m AA_mut \u001b[38;5;241m=\u001b[39m \u001b[43mmut_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mAA_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mMut_list\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     87\u001b[0m AA_muts \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(AA_mut)\n\u001b[0;32m     88\u001b[0m MUT_sum, WT_diff \u001b[38;5;241m=\u001b[39m diff_weighted(pos_corr, neg_corr, AA_muts, ideal_pos_value, ideal_neg_value)\n",
      "File \u001b[1;32mc:\\Users\\marik\\OneDrive - bwedu\\Uni HD\\FS 4\\Bioinfo\\topic04_02\\function_mut.py:59\u001b[0m, in \u001b[0;36mmut_apply\u001b[1;34m(AA_list, Mut_list)\u001b[0m\n\u001b[0;32m     54\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m     55\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m sum_dev, WT_weight \u001b[38;5;66;03m#sum_dev is a positive value of all deviations, higher values indicate a worse fit\u001b[39;00m\n\u001b[1;32m---> 59\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mmut_apply\u001b[39m(AA_list, Mut_list):\n\u001b[0;32m     60\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m     61\u001b[0m \u001b[38;5;124;03m    Applies a list of mutations to a list of amino acids.\u001b[39;00m\n\u001b[0;32m     62\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     68\u001b[0m \u001b[38;5;124;03m        list: A new list containing the amino acid sequence after applying the mutations.\u001b[39;00m\n\u001b[0;32m     69\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m     70\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(Mut_list) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#test_list = ['AF-P0AGD1-F1.pdb', 'AF-O34633-F1.pdb', 'AF-Q72L06-F1.pdb', 'AF-Q746J6-F1.pdb', 'AF-R4YU54-F1.pdb', 'AF-Q72L88-F1.pdb','AF-P21340-F1.pdb' , 'AF-Q745V2-F1.pdb', 'AF-C0H3V2-F1.pdb', 'AF-P39773-F1.pdb', 'AF-R4YTP2-F1.pdb', 'AF-Q6L2N0-F1.pdb', 'AF-P39847-F1.pdb']\n",
    "test_list = ['AF-P0AGD1-F1.pdb', 'AF-O34633-F1.pdb', 'AF-Q72L06-F1.pdb', 'AF-Q746J6-F1.pdb', 'AF-R4YU54-F1.pdb', 'AF-Q72L88-F1.pdb','AF-P21340-F1.pdb' , 'AF-Q745V2-F1.pdb', 'AF-C0H3V2-F1.pdb']\n",
    "results = {}\n",
    "Possible_mut = {}\n",
    "for n in test_list:\n",
    "    pdb_file = n\n",
    "    Dev_sum, Dev, Mut_sum, Mut_dev, Improvement, possible_mut = prot_mut(path_pdb, pdb_file, path_pqr, Deep_mut=True, iterations=5, cutoff_value = 0)\n",
    "    results[n] = Improvement\n",
    "    Possible_mut[n] = possible_mut\n",
    "    print('prot done#########################################################################')\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-0.15281734305166905\n",
      "-0.15281734305166905\n",
      "{'AF-P0AGD1-F1.pdb': 0.0, 'AF-O34633-F1.pdb': 0.0, 'AF-Q72L06-F1.pdb': 0.0, 'AF-Q746J6-F1.pdb': 0.0, 'AF-R4YU54-F1.pdb': 0.0, 'AF-Q72L88-F1.pdb': 0.0, 'AF-P21340-F1.pdb': 0.0, 'AF-Q745V2-F1.pdb': 0.0, 'AF-C0H3V2-F1.pdb': -0.15281734305166905}\n",
      "0.8888888888888888\n",
      "0.08333333333333333\n",
      "0.1349206349206349\n",
      "MRVLFIGDVMAEPGLRAVGLHLPDIRDRYDLVIANGENAARGKGLDRRSYRLLREAGVDLVSLGNHAWDHKEVYALLESEPVVRPLNYPPGTPGKGFWRLEVGGESLLFVQVMGRIFMDPLDDPFRALDRLLEEEKADYVLVEVHAEATSEKMALAHYLDGRASAVLGTHTHVPTLDATRLPKGTLYQTDVGMTGTYHSIIGGEVETFLARFLTGRPQPFRAAQGKARLHATELVFEGGRPVAISPYVWEEP \n",
      "MRVLFIGYVMAEPGLRAVGLHLPYIRYRYYLVIANGENAARGKGLYRRSYRLLREAGVYLVSLGNHAWYHKEVYALLESEPVVRPLNYPPGRPGKGFWRLEVGGESLLFVQVMGRIFMYPLYYPFRALYRLLEEEKAYYVLVEVHAEARSEKMALAHYLYGRASAVLGRHRHVPRLYARRLPKGRLYQRYVGMRGRYHSIIGGEVERFLARFLRGRPQPFRAAQGKARLHARELVFEGGRPVAISPYVWEEP\n"
     ]
    }
   ],
   "source": [
    "'-0.9096700929981782'\n",
    "from function import rel_aa_comp\n",
    "print(re_improvement)\n",
    "print(Improvement_sum)\n",
    "print(results)\n",
    "print(re_count)\n",
    "name = 'AF-Q72L06-F1.pdb'\n",
    "Seq = Mut_seq_dict[name]\n",
    "Wt = AAs_dict[name]\n",
    "Seq = ''.join(Seq)\n",
    "print(rel_aa_comp(Wt, 'R'))\n",
    "print(rel_aa_comp(Seq, 'R'))\n",
    "print(f'{Wt} \\n{Seq}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutator3(AA_list:list, free_AA, deviation, pos_corr:dict, neg_corr:dict, conserv_substitution, ideal_pos_value, ideal_neg_value, cutoff_value, deep_mut=True):\n",
    "    \"\"\"\n",
    "    Generates a list of potential mutations based on deviations and correlations.\n",
    "\n",
    "    Args:\n",
    "        AAs (str): The amino acid sequence.\n",
    "        free_AA (np.ndarray): Array containing information about free amino acids Col1: Prot name, Col2: aminoacid position, Col3: Aminoacid (one letter code).\n",
    "        deviation (list): List containing deviations from ideal values for features.\n",
    "        pos_corr_list (list): List of amino acids that positively correlate with desired features.\n",
    "        sorted_freq_pos (list): Possibly sorted list of frequencies for amino acids contributing to positive features (usage unclear).\n",
    "        neg_corr_list (list): List of amino acids that negatively correlate with desired features.\n",
    "        conserv_substitution (dict): Dictionary containing a list of possible conservative substitutions for each amino acid.\n",
    "\n",
    "    Returns:\n",
    "        list: Mutated protein as a list\n",
    "        list: List of mutations (AA-POS-AA)\n",
    "    \"\"\"\n",
    "    \n",
    "    import itertools\n",
    "    from function import rel_aa_comp\n",
    "    from function_mut import mut_apply\n",
    "    from function_mut import mut_live_test\n",
    "    \n",
    "    pos_corr_list = list(pos_corr.keys())\n",
    "    neg_corr_list = list(neg_corr.keys())\n",
    "    AA_mut_list = AA_list\n",
    "    AAs = ''.join(AA_list)\n",
    "    mut_AAs = AAs\n",
    "    free_AA_dict = {a: b for a, b in zip(free_AA[:,2], free_AA[:,1] )} # create dictionary from array the key is the absolute aminoacid position and value is the aminoacid\n",
    "    mut_list = []\n",
    "    final_Diff_cutoff = cutoff_value\n",
    "    first_entry = deviation[0][0]\n",
    "    max_increase = 1.3 # maximum increase of relative amino acid composition\n",
    "    \n",
    "    \n",
    "    \n",
    "    if deep_mut == True:\n",
    "        for n in range(2):  #goes throught the first 6 highest deviations\n",
    "            first_entry = deviation[n][0]\n",
    "            Diff_start = final_Diff_cutoff + 0.1\n",
    "            count = 0\n",
    "            \n",
    "            while AAs == mut_AAs:\n",
    "                count += 1\n",
    "                Diff_start += -0.01\n",
    "                print(f'current level of deviation: {deviation[n]} -- {first_entry} --- cutoff {Diff_start}')\n",
    "                if Diff_start <= final_Diff_cutoff:\n",
    "                    break\n",
    "                \n",
    "                else:    \n",
    "                    # determine possible substitutions if the first entry is a single amino acid\n",
    "                    if len(first_entry) == 1:\n",
    "                    \n",
    "                        if first_entry in pos_corr_list: #checks if amount of aminoacid should be increaed\n",
    "                            for key in free_AA_dict:\n",
    "                                aminoacid = free_AA_dict[key]\n",
    "                                AA_subst = conserv_substitution[aminoacid]\n",
    "                                \n",
    "                                \n",
    "                                for k in pos_corr_list: # Check if the current amino acid is in the positive correlation list\n",
    "                                    if len(k) == 1:     #selects the first feature with one aminoacid\n",
    "                                        if rel_aa_comp(mut_AAs, k) < ideal_pos_value[k]:    #checks if the relative composition is suboptimal\n",
    "                                            if k in AA_subst:\n",
    "                                                mut_aa = (aminoacid + '-' + key + '-' + k)\n",
    "                                                Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                                if Diff > Diff_start:\n",
    "                                                    mut_list.append(mut_aa)\n",
    "                                                    AA_mut_list = mut_apply(AA_list, [mut_aa]) #applies mutation to the protein sequence\n",
    "                                                    mut_AAs = ''.join(AA_mut_list)  #needs to be adjustet to not change original sequence\n",
    "                                                    break\n",
    "\n",
    "\n",
    "                        elif first_entry in neg_corr_list: #checks if amount of aminoacid should be decreased\n",
    "                            for key in free_AA_dict:\n",
    "                                aminoacid = free_AA_dict[key]\n",
    "                                if aminoacid == first_entry:\n",
    "                                    AA_subst = conserv_substitution[aminoacid]\n",
    "                                    \n",
    "                                    #tries to substitute the aminoacid to the aminoacids that comes first in the sorted_freq_pos list\n",
    "                                    for k in pos_corr_list:\n",
    "                                        if aminoacid == k:\n",
    "                                            if rel_aa_comp(mut_AAs, k) < ideal_pos_value[k]:\n",
    "                                                mut_aa = aminoacid + '-' + key + '-' + k\n",
    "                                                Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                                if Diff > Diff_start: # live test if mutation is benefitical\n",
    "                                                    AA_mut_list = mut_apply(AA_list, [mut_aa]) #applies mutation to the protein sequence\n",
    "                                                    mut_list.append(mut_aa)\n",
    "                                                    mut_AAs = ''.join(AA_mut_list)\n",
    "                                                    break\n",
    "                                        \n",
    "                                        \n",
    "\n",
    "                    # determine possible substitutions if the first entry is a pair of amino acids\n",
    "                    elif len(first_entry) == 2:\n",
    "                        \n",
    "                        if first_entry in pos_corr_list: # checks if the amount of aminoacids should be increased\n",
    "                            for key in free_AA_dict:\n",
    "                                aminoacid = free_AA_dict[key]\n",
    "                                #if aminoacid not in sorted_freq_pos: #checks if the aminoacid overall positively contributes to one of the pos_corr features\n",
    "                                AA_subst = conserv_substitution[aminoacid] # list of possible substitutions for the current amino acid\n",
    "                                \n",
    "                                #mutation\n",
    "                                for n in range(len(deviation)): # iterates through all deviations and takes the first deviation (from pos corr) which can be increaed\n",
    "                                    entry = deviation[n][0]\n",
    "                                    if entry in pos_corr_list:\n",
    "                                        subst = [] #creates a list of possible substitutions that increase one of the amino acids in the highest entry that positively correlates\n",
    "                                        for k in AA_subst:\n",
    "                                            if k in first_entry:\n",
    "                                                subst = k\n",
    "                                                \n",
    "                                        if len(subst) == 1: # if only one substitution increases one of the aminoacids this substitution will be used\n",
    "                                            mut_aa = aminoacid + '-' + key + '-' + subst\n",
    "                                            Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                            mut_rel = rel_aa_comp(AAs, subst)\n",
    "                                            mut_rel_max = rel_aa_comp(AAs, subst) * max_increase\n",
    "                                            if Diff > Diff_start and mut_rel <= mut_rel_max:\n",
    "                                                AA_mut_list = mut_apply(AA_list, [mut_aa]) #applies mutation to the protein sequence\n",
    "                                                mut_list.append(mut_aa)\n",
    "                                                mut_AAs = ''.join(AA_mut_list)\n",
    "                                                break\n",
    "                                            \n",
    "                                        elif len(subst) == 2: # if the aminoacid can be substituted to both aminoacids in first_entry choose the one that has the least frequency\n",
    "                                            comp = [(aa, rel_aa_comp(AAs, aa)) for aa in subst] # calculate the relative amino acid composition of the possible substitutions\n",
    "                                            lowest_comp = min(comp, key=lambda pair: pair[1]) # find the amino acid with the lowest relative composition\n",
    "                                            mut_aa = aminoacid + '-' + key + '-' + lowest_comp[0]\n",
    "                                            mut_rel = rel_aa_comp(AAs, lowest_comp[0])\n",
    "                                            mut_rel_max = rel_aa_comp(AAs, lowest_comp[0]) * max_increase\n",
    "                                            Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                            if Diff > Diff_start and mut_rel <= mut_rel_max:\n",
    "                                                AA_mut_list = mut_apply(AA_list, [mut_aa])\n",
    "                                                mut_list.append(mut_aa) # append the mutation to the mutation list\n",
    "                                                mut_AAs = ''.join(AA_mut_list)\n",
    "                                                break\n",
    "\n",
    "                        \n",
    "                                    \n",
    "                        elif first_entry in neg_corr_list: # checks if the amount of aminoacids should be decreased\n",
    "                            for key in free_AA_dict:\n",
    "                                aminoacid = free_AA_dict[key]\n",
    "                                if aminoacid in first_entry: #checks if the aminoacid is present in the first_entry\n",
    "                                    AA_subst = conserv_substitution[aminoacid] # list of possible substitutions for the current amino acid\n",
    "                                    \n",
    "                                    #mutation               \n",
    "                                    for n in range(len(deviation)): # iterates through all deviations and takes the first deviation (from pos corr) which can be increaed\n",
    "                                        entry = deviation[n][0]\n",
    "                                        if entry in pos_corr_list: \n",
    "                                            subst = [] #creates a list of possible substitutions that increase one of the amino acids in the first_entry\n",
    "                                            for k in AA_subst:\n",
    "                                                if k in entry:\n",
    "                                                    subst.append(k)\n",
    "                                            if len(subst) == 1: # if only one substitution increases one of the aminoacids this substitution will be used\n",
    "                                                mut_aa = aminoacid + '-' + key + '-' + subst[0]\n",
    "                                                mut_rel = rel_aa_comp(mut_AAs, subst[0])\n",
    "                                                mut_rel_max = rel_aa_comp(AAs, subst[0]) * max_increase\n",
    "                                                Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                                if Diff > Diff_start and mut_rel <= mut_rel_max:\n",
    "                                                    AA_mut_list = mut_apply(AA_list, [mut_aa])\n",
    "                                                    mut_list.append(mut_aa)\n",
    "                                                    mut_AAs = ''.join(AA_mut_list) \n",
    "                                                    break\n",
    "                                                \n",
    "                                            elif len(subst) == 2: # if the aminoacid can be substituted to both aminoacids in first_entry choose the one that has the least frequency\n",
    "                                                comp = [(aa, rel_aa_comp(mut_AAs, aa)) for aa in subst] # calculate the relative amino acid composition of the possible substitutions\n",
    "                                                lowest_comp = min(comp, key=lambda pair: pair[1]) # find the amino acid with the lowest relative composition\n",
    "                                                mut_aa = aminoacid + '-' + key + '-' + lowest_comp[0]\n",
    "                                                \n",
    "                                                Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                                mut_rel = rel_aa_comp(mut_AAs, lowest_comp[0])\n",
    "                                                mut_rel_max = rel_aa_comp(AAs, lowest_comp[0]) * max_increase\n",
    "                                                \n",
    "                                                if Diff > Diff_start and mut_rel <= mut_rel_max:\n",
    "                                                    AA_mut_list = mut_apply(AA_list, [mut_aa])\n",
    "                                                    mut_list.append(mut_aa) # append the mutation to the mutation list\n",
    "                                                    mut_AAs = ''.join(AA_mut_list)\n",
    "                                                    break\n",
    "                    \n",
    "\n",
    "                    elif 'motif' in first_entry:\n",
    "                        if first_entry in pos_corr_list: # checks if the amount of aminoacids should be increased\n",
    "                            for key in free_AA_dict:\n",
    "                                prev_key = int(key) -2\n",
    "                                curr_key = int(key) -1\n",
    "                                for_key = int(key) \n",
    "                                \n",
    "                                \n",
    "                                #handles edge-case for first aminoacid in the sequence\n",
    "                                if key == 1: #if the aminoacid is the first in the sequence\n",
    "                                    curr_aa = AAs[curr_key]\n",
    "                                    for_aa = AAs[for_key]\n",
    "                                                                                    \n",
    "                                    if curr_aa not in first_entry[0] and for_aa in first_entry[1]:\n",
    "                                        curr_subst = conserv_substitution[curr_aa]\n",
    "                                        for s in curr_subst:\n",
    "                                            if s in first_entry[0]:\n",
    "                                                s_rel = rel_aa_comp(mut_AAs, s)\n",
    "                                                s_rel_max = rel_aa_comp(AAs, s) * max_increase\n",
    "                                                if s in first_entry[0] and s_rel <= s_rel_max:\n",
    "                                                    mut_aa = f'{curr_key}-{for_key}-{s}'\n",
    "                                                    Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value)\n",
    "                                                    if Diff > Diff_start:\n",
    "                                                        AA_mut_list = mut_apply(AA_list, [mut_aa])\n",
    "                                                        mut_list.append(mut_aa)\n",
    "                                                        mut_AAs = ''.join(AA_mut_list)\n",
    "                                                        break\n",
    "                                \n",
    "                                \n",
    "                                #handles edge-case if for the last aminoacid in the sequence\n",
    "                                elif for_key >= len(AAs):  #if the aminoacid is the last in the sequence\n",
    "                                    prev_aa = AAs[prev_key]\n",
    "                                    curr_aa = AAs[curr_key]\n",
    "                                    if prev_aa not in first_entry[0] or curr_aa not in first_entry[1]:\n",
    "                                        \n",
    "                                        if prev_aa in first_entry[0] and curr_aa not in first_entry[1]:\n",
    "                                            curr_subst = conserv_substitution[curr_aa]\n",
    "                                            for s in curr_subst:\n",
    "                                                if s in first_entry[1]:\n",
    "                                                    s_rel = rel_aa_comp(mut_AAs, s)\n",
    "                                                    s_rel_max = rel_aa_comp(AAs, s) * max_increase\n",
    "                                                    if s in first_entry[1] and s_rel <= s_rel_max:\n",
    "                                                        mut_aa = f'{prev_key}-{curr_key}-{s}'\n",
    "                                                        Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value)\n",
    "                                                        if Diff > Diff_start:\n",
    "                                                            AA_mut_list = mut_apply(AA_list, [mut_aa])\n",
    "                                                            mut_list.append(mut_aa)\n",
    "                                                            mut_AAs = ''.join(AA_mut_list)\n",
    "                                                            break\n",
    "                                \n",
    "                                \n",
    "                                else:            \n",
    "                                    aa_back = AAs[prev_key]    # aminoacid before the current aminoacid\n",
    "                                    aa_current = AAs[curr_key]   # current aminoacid\n",
    "                                    aa_for = AAs[for_key]     # aminoacid after the current aminoacid\n",
    "                                    \n",
    "                                    if aa_back in first_entry[0] and aa_current not in first_entry[1]: #checks if previous aminoacid is part of the motif, and current aminoacid not\n",
    "                                        curr_subst = conserv_substitution[aa_current] # possible substitutions for the current aminoacid\n",
    "                                        for s in curr_subst:\n",
    "                                            s_rel = rel_aa_comp(mut_AAs, s)\n",
    "                                            s_rel_max = rel_aa_comp(AAs, s) * max_increase\n",
    "                                            if s in first_entry[1] and s_rel <= s_rel_max:\n",
    "                                                mut_aa = f'{curr_key}-{for_key}-{s}'\n",
    "                                                Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                                if Diff > Diff_start:\n",
    "                                                    AA_mut_list = mut_apply(AA_list, [mut_aa])\n",
    "                                                    mut_list.append(mut_aa)\n",
    "                                                    mut_AAs = ''.join(AA_mut_list)\n",
    "                                                    break\n",
    "                                    \n",
    "                                    if aa_current not in first_entry[0] and aa_for in first_entry[1]: #checks if current aminoacid is not part of the motif, and next aminoacid is\n",
    "                                        curr_subst = conserv_substitution[aa_current] # possible substitutions for the current aminoacid\n",
    "                                        for s in curr_subst:\n",
    "                                            s_rel = rel_aa_comp(mut_AAs, s)\n",
    "                                            s_rel_max = rel_aa_comp(AAs, s) * max_increase\n",
    "                                            if s in first_entry[0] and s_rel <= s_rel_max:\n",
    "                                                mut_aa = f'{curr_key}-{for_key}-{s}'\n",
    "                                                Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value)\n",
    "                                                if Diff > Diff_start:\n",
    "                                                    AA_mut_list = mut_apply(AA_list, [mut_aa])\n",
    "                                                    mut_list.append(mut_aa)\n",
    "                                                    mut_AAs = ''.join(AA_mut_list)\n",
    "                                                    break\n",
    "            if AAs != mut_AAs:\n",
    "                break\n",
    "                            \n",
    "    return AA_mut_list, mut_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10  1  2\n"
     ]
    }
   ],
   "source": [
    "key = 1\n",
    "prev_key = key-2\n",
    "for_key = key\n",
    "curr_key = key-1\n",
    "test = [1,2,3,4,5,6,7,8,9,10]\n",
    "print(f'{test[prev_key]}  {test[curr_key]}  {test[for_key]}')\n",
    "#print(len(test))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get list of pos/neg correlating features for Meltpoint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\marik\\AppData\\Local\\Temp\\ipykernel_16368\\768696474.py:3: DtypeWarning: Columns (8,21,22,23,25,26,27,28) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  pro_df: pd.DataFrame = pd.read_csv(os.path.join('./data', 'prokaryotes_323columns.csv'))\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "pro_df: pd.DataFrame = pd.read_csv(os.path.join('./data', 'prokaryotes_323columns.csv'))\n",
    "pro_df = pro_df.drop(columns = ['Unnamed: 0', 'Unnamed: 0.1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ARmotif': 0.010758762261349177, 'EAmotif': 0.01633251571740483}\n",
      "{}\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "top_df = pro_df[pro_df['meltPoint'] >= pro_df['meltPoint'].quantile(0.9)]\n",
    "AR_mean = np.mean(top_df['AR'])\n",
    "Ideal_pos_value = {}\n",
    "Ideal_neg_value = {}\n",
    "\n",
    "keys = ['ARmotif', 'EAmotif']\n",
    "for key in keys:\n",
    "    Ideal_pos_value[key] = np.mean(top_df[key])\n",
    "\n",
    "\"\"\" for key in neg_corr.keys():\n",
    "    Ideal_neg_value[key] = np.mean(top_df[key]) \"\"\"\n",
    "    \n",
    "print(Ideal_pos_value)\n",
    "print(Ideal_neg_value)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "pro_corr = pro_df.corr(numeric_only=True)['meltPoint']\n",
    "pos_corr = pro_corr[pro_corr > 0.32]\n",
    "neg_corr = pro_corr[pro_corr < -0.32]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['IQ', 'MQ', 'MT', 'FQ', 'WQ', 'NQ', 'QS', 'QT', 'QD', 'QH', 'QK', 'QC', 'ST', 'SH', 'SC', 'TD', 'TH', 'TC', 'Qhelix', 'Q', 'T', 'PolarAA']\n"
     ]
    }
   ],
   "source": [
    "pos_corr = list(pro_corr.index)\n",
    "neg_corr = list(neg_corr.index)\n",
    "print(neg_corr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Needs to be mutated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ        -0.429683\n",
      "MQ        -0.507329\n",
      "MT        -0.359458\n",
      "FQ        -0.423630\n",
      "WQ        -0.420057\n",
      "NQ        -0.435562\n",
      "QS        -0.502697\n",
      "QT        -0.529350\n",
      "QD        -0.466556\n",
      "QH        -0.455041\n",
      "QK        -0.418720\n",
      "QC        -0.493738\n",
      "ST        -0.396185\n",
      "SH        -0.327684\n",
      "SC        -0.321686\n",
      "TD        -0.344690\n",
      "TH        -0.346921\n",
      "TC        -0.364763\n",
      "Qhelix    -0.378950\n",
      "Q         -0.469765\n",
      "T         -0.321208\n",
      "PolarAA   -0.406035\n",
      "Name: meltPoint, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "print(pro_corr[pro_corr < -0.32])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
