{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Working jupyter to create function to create a list of aminoacids likely to be used for mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1000\n",
      "ATOM   7949  N   ASN A 1000     -16.817  14.355  -9.711  1.00 97.93           N\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "line = 'ATOM   7949  N   ASN A1000     -16.817  14.355  -9.711  1.00 97.93           N'\n",
    "line = re.sub(r'([A-Za-z])(\\d)', r'\\1 \\2', line)\n",
    "print(line.split()[5])\n",
    "print(line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/pdbs\\AF-C0H3V2-F1.pdb\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "path_pdb = './data/pdbs'\n",
    "path_pqr=  './data/pqrs'\n",
    "pdb_file = 'AF-C0H3V2-F1.pdb'\n",
    "output_path = './data/test'\n",
    "\n",
    "print(os.path.join(path_pdb, pdb_file))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fasparse(faspath):\n",
    "    import numpy as np\n",
    "    helixl = []\n",
    "    sheetl = []\n",
    "    data = np.loadtxt(faspath, dtype={'names': ('index', 'col1', 'col2', 'val1', 'val2', 'val3'),'formats': ('i4', 'S1', 'S1', 'f4', 'f4', 'f4')})\n",
    "    indiceshelix = data['index'][data['col2'] == b'H']\n",
    "    indicessheet = data['index'][data['col2'] == b'E']\n",
    "    helix = np.split(indiceshelix, np.where(np.diff(indiceshelix) != 1)[0]+1)\n",
    "    sheet = np.split(indicessheet, np.where(np.diff(indicessheet) != 1)[0]+1)\n",
    "    return [helix, sheet]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "def AA2s4pred (directory_S4pred, output_path, AA_seq, prot):\n",
    "    import os\n",
    "    os.getcwd()\n",
    "    # call s4pred and create fas file\n",
    "    fastapath = os.path.join(output_path,f'{prot}.fasta')\n",
    "    faspath = os.path.join(output_path,f'{prot}.fas')\n",
    "    \n",
    "    if os.path.isfile(fastapath):\n",
    "        print(f'fasta file already exists {fastapath}')\n",
    "    else:\n",
    "        with open(os.path.join(output_path,f'{prot}.fasta'), \"w\") as fasta_file:\n",
    "            fasta_file.write(f\">{prot}\\n{AA_seq}\\n\")\n",
    "        print(f'fasta file created {fastapath}')\n",
    "    if os.path.isfile(faspath):\n",
    "        print('fas file already exists')\n",
    "    else:\n",
    "        os.chdir(directory_S4pred)\n",
    "        os.system(f'python3 run_model.py \"{fastapath}\" > \"{faspath}\"')\n",
    "        os.chdir('../../')\n",
    "     \n",
    "    #read fas file and   \n",
    "    sec_pred = fasparse(faspath)    \n",
    "\n",
    "    return sec_pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n"
     ]
    }
   ],
   "source": [
    "for i in range(10):\n",
    "    if i >= 0:\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prot_mut(pdb_path, pdb_file, pqr_output_path, deep_mut=None, iterations=1):\n",
    "    \n",
    "    import os\n",
    "    from helper_function import pdb2AA\n",
    "    from function import functional_aa\n",
    "    from function import free_aa\n",
    "    import numpy as np\n",
    "    \n",
    "    \n",
    "    # top 20 features that correlate with melt point (positive and negative), based on prokaryotes323 columns, cutoff at +/- 0.32\n",
    "    pos_corr = {'YR': 0.492442, \n",
    "                'RP': 0.480594, \n",
    "                'RG': 0.442533, \n",
    "                'R': 0.432905, \n",
    "                'WR': 0.392523, \n",
    "                'YP': 0.390709, \n",
    "                'LR': 0.386596, \n",
    "                'FR': 0.376553, \n",
    "                'VR': 0.370241, \n",
    "                'ER': 0.369174, \n",
    "                'RC': 0.357052, \n",
    "                #'Rhelix': 0.349204, # percentage of R in all helices\n",
    "                'MR': 0.343496, \n",
    "                'P': 0.340082, \n",
    "                'PG': 0.338434, \n",
    "                'EAmotif': 0.332202, # EA hinterenander\n",
    "                'LP': 0.330337, \n",
    "                'EP': 0.328569, \n",
    "                'RH': 0.326193, \n",
    "                'AR': 0.324355, \n",
    "                'ARmotif': 0.3241, # AR hinterenander\n",
    "                'NR': 0.323378}\n",
    "    \n",
    "    neg_corr = {'QT': -0.52935, \n",
    "                'MQ': -0.507329, \n",
    "                'QS': -0.502697, \n",
    "                'QC': -0.493738, \n",
    "                'Q': -0.469765, \n",
    "                'QD': -0.466556, \n",
    "                'QH': -0.455041, \n",
    "                'NQ': -0.435562, \n",
    "                'IQ': -0.429683, \n",
    "                'FQ': -0.42363, \n",
    "                'WQ': -0.420057, \n",
    "                'QK': -0.41872, \n",
    "                #'PolarAA': -0.406035, \n",
    "                'ST': -0.396185, \n",
    "                #'Qhelix': -0.37895, \n",
    "                'TC': -0.364763, \n",
    "                'MT': -0.359458, \n",
    "                'TH': -0.346921, \n",
    "                'TD': -0.34469, \n",
    "                'SH': -0.327684, \n",
    "                'SC': -0.321686, \n",
    "                'T': -0.321208}\n",
    "    sorted_freq_pos = ['R', 'P', 'L', 'Y', 'E', 'G', 'A', 'V', 'M', 'F', 'W', 'N', 'H', 'C']\n",
    "    sorted_freq_neg = ['K', 'N', 'W', 'F', 'I', 'D', 'M', 'C', 'H', 'S', 'T', 'Q']\n",
    "\n",
    "    \n",
    "    ideal_pos_value = {'AR': 0.19577028919092312, 'VR': 0.16570774850840858, 'LR': 0.22057443760531548, 'LP': 0.19913083779456836, 'MR': 0.10189569954318338, 'FR': 0.1214807452476434, 'WR': 0.09732041140609597, 'NR': 0.10350733491352844, 'YR': 0.11505796446768121, 'YP': 0.09361436465693411, 'ER': 0.1750096588731577, 'EP': 0.15356605906241058, 'RH': 0.10505828276892132, 'RC': 0.0895071066024632, 'RP': 0.14867680267495723, 'RG': 0.1743851008086472, 'PG': 0.15294150099790005, 'R': 0.08506020124285214, 'P': 0.06361660143210504}\n",
    "    ideal_neg_value = {'IQ': 0.05490211183163266, 'MQ': 0.03973022298244349, 'MT': 0.054568274035922376, 'FQ': 0.05931526868690351, 'WQ': 0.03515493484535609, 'NQ': 0.041341858352788544, 'QS': 0.057844351285310555, 'QT': 0.06062750041770345, 'QD': 0.06216042713743332, 'QH': 0.04289280620818142, 'QK': 0.06281534035966867, 'QC': 0.027341630041723304, 'ST': 0.07268240233878943, 'SH': 0.054947708129267414, 'SC': 0.03939653196280929, 'TD': 0.0769984781909122, 'TH': 0.0577308572616603, 'TC': 0.04217968109520218, 'Q': 0.022894724682112268, 'T': 0.03773277573559115}\n",
    "\n",
    "\n",
    "    #https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003674\n",
    "    conserv_subst = {\n",
    "        'A': ['D', 'E', 'G', 'S', 'T'],\n",
    "        'C': ['G', 'R', 'S', 'W', 'Y'],\n",
    "        'D': ['A', 'E', 'G', 'H', 'N', 'V', 'Y'],\n",
    "        'E': ['A', 'D', 'G', 'K', 'Q', 'V'],\n",
    "        'F': ['I', 'L', 'Y'],\n",
    "        'G': ['A', 'C', 'D', 'E', 'R'],\n",
    "        'H': ['D', 'L', 'N', 'P', 'Q', 'R', 'Y'],\n",
    "        'I': ['F', 'L', 'M', 'N', 'V'],\n",
    "        'K': ['E', 'M', 'N', 'Q', 'R', 'T'],\n",
    "        'L': ['F', 'H', 'I', 'M', 'P', 'Q', 'R', 'V', 'W'],\n",
    "        'M': ['I', 'K', 'L', 'R', 'T', 'V'],\n",
    "        'N': ['D', 'H', 'I', 'K', 'S', 'T', 'Y'],\n",
    "        'P': ['H', 'L', 'Q', 'R', 'S'],\n",
    "        'Q': ['E', 'H', 'K', 'L', 'P', 'R'],\n",
    "        'R': ['C', 'G', 'H', 'K', 'L', 'M', 'P', 'Q', 'T', 'W'],\n",
    "        'S': ['A', 'C', 'N', 'P', 'T', 'W', 'Y'],\n",
    "        'T': ['A', 'K', 'M', 'N', 'R', 'S'],\n",
    "        'V': ['D', 'E', 'I', 'L', 'M'],\n",
    "        'W': ['C', 'L', 'R', 'S'],\n",
    "        'Y': ['C', 'D', 'F', 'H', 'N'],\n",
    "        }\n",
    "\n",
    "    non_conservative_substitutions = {\n",
    "        'A': ['P', 'V'],\n",
    "        'C': ['F'],\n",
    "        'F': ['C', 'S', 'V'],\n",
    "        'G': ['S', 'V', 'W'],\n",
    "        'I': ['K', 'R', 'S', 'T'],\n",
    "        'K': ['I'],\n",
    "        'L': ['S'],\n",
    "        'P': ['A', 'T'],\n",
    "        'Q': ['E', 'K'],\n",
    "        'R': ['I', 'S'],\n",
    "        'S': ['F', 'G', 'I', 'L', 'R'],\n",
    "        'T': ['I', 'P'],\n",
    "        'V': ['A', 'F', 'G'],\n",
    "        'W': ['G'],\n",
    "        }\n",
    "\n",
    "    #extract aminoacid list\n",
    "    aa_list = pdb2AA(pdb_path, pdb_file)\n",
    "    aa_locked = functional_aa(pdb_path, pdb_file, pqr_output_path)\n",
    "    aa_free = free_aa(pdb_path, pdb_file, aa_locked)\n",
    "    AAs = ''.join(aa_list)\n",
    "    # calculate ratio of WT protein for pos_corr and neg_corr and sort them in descending order\n",
    "    from function import rel_aa_comp\n",
    "    from function_mut import diff_weighted\n",
    "    from function_mut import mut_apply\n",
    "    from function_mut import mutator2\n",
    "    from helper_function import ArraySlice\n",
    "    WT_dev_sum, WT_dev = diff_weighted(pos_corr, neg_corr, aa_list, ideal_pos_value, ideal_neg_value)\n",
    "    \n",
    "    #define variables for iteration\n",
    "    cutoff = 0.01\n",
    "    prev_Mut_list = aa_list\n",
    "    prev_Mut_dev_sum = WT_dev_sum\n",
    "    prev_Mut_dev = WT_dev\n",
    "    aa_available = aa_free\n",
    "    \n",
    "    Dev_list =[WT_dev]\n",
    "    \n",
    "    for i in range(iterations):\n",
    "        Mut_list, possible_mutations = mutator2(prev_Mut_list, aa_available, prev_Mut_dev ,pos_corr, sorted_freq_pos, neg_corr, conserv_subst, ideal_pos_value, ideal_neg_value) # calculates list of possible mutations\n",
    "        Mut_dev_sum, Mut_dev = diff_weighted(pos_corr, neg_corr, Mut_list, ideal_pos_value, ideal_neg_value) # calculate deviation of mutated protein sequence\n",
    "        Dev_list.append(Mut_dev)\n",
    "        if Mut_dev_sum - cutoff >= prev_Mut_dev_sum:\n",
    "           output_message = f' After {i+1} iterations: Mutation increased deviation of protein sequence from ideal values \\n{WT_dev_sum} \\n {prev_Mut_dev_sum} : {i} iterations -> {Mut_dev_sum} : {i+1} iterations'\n",
    "           break\n",
    "        elif prev_Mut_dev[0][0] == Mut_dev[0][0]:\n",
    "            output_message = f' After {i+1} no change in the most deviating feature'\n",
    "            break\n",
    "        \n",
    "        #update variables for next iteration            \n",
    "        prev_possible_mutations = possible_mutations\n",
    "        prev_Mut_list = Mut_list\n",
    "        prev_Mut_dev = Mut_dev\n",
    "        prev_Mut_dev_sum = Mut_dev_sum\n",
    "        aa_available = ArraySlice(aa_available, possible_mutations) #updates available aminoacids, so that each aminoacid can only be mutated once\n",
    "        if i == (iterations-1):\n",
    "            output_message = f'maximum amount of iterations reached: {i+1} \\n deviation from {WT_dev_sum},  to {Mut_dev_sum}'\n",
    "\n",
    "    \n",
    "\n",
    "    #introduce random mutations\n",
    "    if i < iterations:\n",
    "        R_prev_Mut_list, R_possible_mutations = random_mutate(prev_Mut_list, aa_available, conserv_subst, 0.01)\n",
    "        \n",
    "        R_possible_mutations = possible_mutations\n",
    "#        R_Mut_dev = Mut_dev\n",
    "        R_aa_available = aa_available\n",
    "        R_prev_Mut_dev = prev_Mut_dev\n",
    "        R_prev_Mut_dev_sum = prev_Mut_dev_sum\n",
    "        \n",
    "        for i in range(iterations):\n",
    "            R_Mut_list, R_possible_mutations = mutator2(R_prev_Mut_list, R_aa_available, R_prev_Mut_dev ,pos_corr, sorted_freq_pos, neg_corr, conserv_subst, ideal_pos_value, ideal_neg_value) # calculates list of possible mutations\n",
    "            R_Mut_dev_sum, R_Mut_dev = diff_weighted(pos_corr, neg_corr, R_Mut_list, ideal_pos_value, ideal_neg_value) # calculate deviation of mutated protein sequence\n",
    "            Dev_list.append(R_Mut_dev)\n",
    "            \n",
    "            if i > 0:\n",
    "                if R_Mut_dev_sum - cutoff >= R_prev_Mut_dev_sum:\n",
    "                    output_message = f'RAND After {i+1} iterations: Mutation increased deviation of protein sequence from ideal values \\n{WT_dev_sum} \\n {R_prev_Mut_dev_sum} : {i} iterations -> {R_Mut_dev_sum} : {i+1} iterations'\n",
    "                    break\n",
    "                #elif R_prev_Mut_dev[0][0] == R_Mut_dev[0][0]:\n",
    "                #   output_message = f'RAND After {i+1} no change in the most deviating feature'\n",
    "                #  break\n",
    "            \n",
    "            #update variables for next iteration\n",
    "            R_prev_possible_mutations = R_possible_mutations\n",
    "            R_prev_Mut_list = R_Mut_list\n",
    "            R_prev_Mut_dev = R_Mut_dev\n",
    "            R_prev_Mut_dev_sum = R_Mut_dev_sum\n",
    "            R_aa_available = ArraySlice(R_aa_available, R_possible_mutations) #updates available aminoacids, so that each aminoacid can only be mutated once\n",
    "            if i == (iterations-1):\n",
    "                output_message = f'RAND maximum amount of iterations reached: {i+1} \\n deviation from {WT_dev_sum},  to {Mut_dev_sum}'\n",
    "            \n",
    "    \n",
    "    print( f'########## \\n {output_message} \\n ##########')  \n",
    "    \n",
    "    if R_prev_Mut_dev_sum < prev_Mut_dev_sum:\n",
    "        R_Improvement = R_prev_Mut_dev_sum - WT_dev_sum\n",
    "        return R_Improvement, Dev_list\n",
    "    else:\n",
    "        print('no improvement with random mutations')\n",
    "        Improvement = prev_Mut_dev_sum - WT_dev_sum\n",
    "        return Improvement, Dev_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "def random_mutate(AA_list, free_AA, mutation_dict, mutation_rate):\n",
    "    \"\"\"\n",
    "    Mutates a protein sequence randomly, respecting allowed mutations from a dictionary.\n",
    "\n",
    "    Args:\n",
    "        sequence (str): The protein sequence to mutate.\n",
    "        mutation_dict (dict): Dictionary where keys are amino acids and values are lists of possible mutations.\n",
    "\n",
    "    Returns:\n",
    "        str: The mutated protein sequence.\n",
    "    \"\"\"\n",
    "    import random\n",
    "    from function_mut import mut_apply\n",
    "    mut_list = []\n",
    "    aa_mut_list = AA_list\n",
    "    free_AA_dict = {a: b for a, b in zip(free_AA[:,2], free_AA[:,1] )} # create dictionary from array the key is the absolute aminoacid position and the value the aminoacid\n",
    "\n",
    "    for key in free_AA_dict:\n",
    "        aminoacid = key\n",
    "        if aminoacid in mutation_dict:\n",
    "            possible_mutation = mutation_dict[aminoacid]\n",
    "            if random.random() < mutation_rate:\n",
    "                mutation = random.choice(possible_mutation)\n",
    "                mut_aa = f'{aminoacid} - {free_AA_dict[key]} - {mutation}'\n",
    "                mut_list.append(mut_aa)\n",
    "                aa_mut_list = mut_apply(AA_list, [mut_aa])\n",
    "    return aa_mut_list, mut_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pqr file already exists\n",
      "Salt_bridge finished\n",
      "H_bond finished\n",
      "VdW_interaction finished\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\marik\\OneDrive - bwedu\\Uni HD\\FS 4\\Bioinfo\\topic04_02\\helper_function.py:71: RuntimeWarning: invalid value encountered in cast\n",
      "  distance[:,0] = distance[:,0].astype('int')\n",
      "c:\\Users\\marik\\OneDrive - bwedu\\Uni HD\\FS 4\\Bioinfo\\topic04_02\\helper_function.py:89: RuntimeWarning: invalid value encountered in arccos\n",
      "  theta = np.arccos((d_DH[n,1,1]**2 + d_HA[n,1:]**2 - d_DA[1:,1:][n,:]**2)/(2*d_DH[n,1,1]*d_HA[1:,1:][n,:]))\n",
      "c:\\Users\\marik\\OneDrive - bwedu\\Uni HD\\FS 4\\Bioinfo\\topic04_02\\helper_function.py:106: RuntimeWarning: invalid value encountered in cast\n",
      "  array[:,0] = array[:,0].astype('int')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "########## \n",
      " RAND maximum amount of iterations reached: 10 \n",
      " deviation from 2.406399223985011,  to 5.062299068488399 \n",
      " ##########\n",
      "no improvement with random mutations\n"
     ]
    }
   ],
   "source": [
    "#test_list = ['AF-P0AGD1-F1.pdb', 'AF-O34633-F1.pdb', 'AF-Q72L06-F1.pdb', 'AF-Q746J6-F1.pdb', 'AF-R4YU54-F1.pdb', 'AF-C0H3V2-F1.pdb']\n",
    "test_list = ['AF-C0H3V2-F1.pdb']\n",
    "results = {}\n",
    "Improvement_sum = 0\n",
    "for n in test_list:\n",
    "    pdb_file = n\n",
    "    imporvement, DEV_list = prot_mut(path_pdb, pdb_file, path_pqr, deep_mut=None, iterations=10)\n",
    "    results[n] = imporvement\n",
    "    Improvement_sum += imporvement\n",
    "\n",
    "re_improvement = Improvement_sum/len(test_list)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-0.2607254885162451\n",
      "-0.2607254885162451\n",
      "{'AF-C0H3V2-F1.pdb': -0.2607254885162451}\n"
     ]
    }
   ],
   "source": [
    "print(re_improvement)\n",
    "print(Improvement_sum)\n",
    "print(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutator3(AA_list:list, free_AA, deviation,pos_corr:dict, sorted_freq_pos, neg_corr:dict, conserv_substitution, ideal_pos_value, ideal_neg_value):\n",
    "    \"\"\"\n",
    "    Generates a list of potential mutations based on deviations and correlations.\n",
    "\n",
    "    Args:\n",
    "        AAs (str): The amino acid sequence.\n",
    "        free_AA (np.ndarray): Array containing information about free amino acids Col1: Prot name, Col2: aminoacid position, Col3: Aminoacid (one letter code).\n",
    "        deviation (list): List containing deviations from ideal values for features.\n",
    "        pos_corr_list (list): List of amino acids that positively correlate with desired features.\n",
    "        sorted_freq_pos (list): Possibly sorted list of frequencies for amino acids contributing to positive features (usage unclear).\n",
    "        neg_corr_list (list): List of amino acids that negatively correlate with desired features.\n",
    "        conserv_substitution (dict): Dictionary containing a list of possible conservative substitutions for each amino acid.\n",
    "\n",
    "    Returns:\n",
    "        list: A list of mutation strings defining potential substitutions.\n",
    "    \"\"\"\n",
    "    \n",
    "    import itertools\n",
    "    from function import rel_aa_comp\n",
    "    from function_mut import mut_apply\n",
    "    from function_mut import mut_live_test\n",
    "    \n",
    "    pos_corr_list = list(pos_corr.keys())\n",
    "    neg_corr_list = list(neg_corr.keys())\n",
    "    AA_mut_list = AA_list\n",
    "    AAs = ''.join(AA_list)\n",
    "    free_AA_dict = {a: b for a, b in zip(free_AA[:,2], free_AA[:,1] )} # create dictionary from array the key is the absolute aminoacid position and the value the aminoacid\n",
    "    mut_list = []\n",
    "    Diff_cutoff = 0\n",
    "    \n",
    "    for n in range(1):\n",
    "        first_entry = deviation[n][0]\n",
    "        # determine possible substitutions if the first entry is a single amino acid\n",
    "        if len(first_entry) == 1:\n",
    "            \n",
    "            if first_entry in pos_corr_list: #checks if amount of aminoacid should be increaed\n",
    "                for key in free_AA_dict:\n",
    "                    aminoacid = free_AA_dict[key]\n",
    "                    AA_subst = conserv_substitution[aminoacid]\n",
    "                    \n",
    "                    \n",
    "                    for k in pos_corr_list: # Check if the current amino acid is in the positive correlation list\n",
    "                        if len(k) == 1:     #selects the first feature with one aminoacid\n",
    "                            if rel_aa_comp(AAs, k) < ideal_pos_value[k]:    #checks if the relative composition is suboptimal\n",
    "                                if k in AA_subst:\n",
    "                                    mut_aa = (aminoacid + '-' + key + '-' + k)\n",
    "                                    Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                    if Diff > Diff_cutoff:\n",
    "                                        mut_list.append(mut_aa)\n",
    "                                        AA_mut_list = mut_apply(AA_list, [mut_aa]) #applies mutation to the protein sequence\n",
    "                                        break\n",
    "\n",
    "\n",
    "            elif first_entry in neg_corr_list: #checks if amount of aminoacid should be decreased\n",
    "                for key in free_AA_dict:\n",
    "                    aminoacid = free_AA_dict[key]\n",
    "                    if aminoacid == first_entry:\n",
    "                        AA_subst = conserv_substitution[aminoacid]\n",
    "                        \n",
    "                        #tries to substitute the aminoacid to the aminoacids that comes first in the sorted_freq_pos list\n",
    "                        for k in pos_corr_list:\n",
    "                            if aminoacid == k:\n",
    "                                if rel_aa_comp(AAs, k) < ideal_pos_value[k]:\n",
    "                                    mut_aa = aminoacid + '-' + key + '-' + k\n",
    "                                    Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                    if Diff > Diff_cutoff: # live test if mutation is benefitical\n",
    "                                        AA_mut_list = mut_apply(AA_list, [mut_aa]) #applies mutation to the protein sequence\n",
    "                                        mut_list.append(mut_aa)\n",
    "                                        break\n",
    "                                \n",
    "                                \n",
    "    \n",
    "        # determine possible substitutions if the first entry is a pair of amino acids\n",
    "        elif len(first_entry) == 2:\n",
    "            \n",
    "            if first_entry in pos_corr_list: # checks if the amount of aminoacids should be increased\n",
    "                for key in free_AA_dict:\n",
    "                    aminoacid = free_AA_dict[key]\n",
    "                    #if aminoacid not in sorted_freq_pos: #checks if the aminoacid overall positively contributes to one of the pos_corr features\n",
    "                    AA_subst = conserv_substitution[aminoacid] # list of possible substitutions for the current amino acid\n",
    "                    \n",
    "                    #mutation\n",
    "                    for n in range(len(deviation)): # iterates through all deviations and takes the first deviation (from pos corr) which can be increaed\n",
    "                        entry = deviation[n][0]\n",
    "                        if entry in pos_corr_list:\n",
    "                            subst = [] #creates a list of possible substitutions that increase one of the amino acids in the highest entry that positively correlates\n",
    "                            for k in AA_subst:\n",
    "                                if k in first_entry:\n",
    "                                    subst = k\n",
    "                                    \n",
    "                            if len(subst) == 1: # if only one substitution increases one of the aminoacids this substitution will be used\n",
    "                                mut_aa = aminoacid + '-' + key + '-' + subst\n",
    "                                Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                if Diff > Diff_cutoff:\n",
    "                                    AA_mut_list = mut_apply(AA_list, [mut_aa]) #applies mutation to the protein sequence\n",
    "                                    mut_list.append(mut_aa)\n",
    "                                    break\n",
    "                                \n",
    "                            elif len(subst) == 2: # if the aminoacid can be substituted to both aminoacids in first_entry choose the one that has the least frequency\n",
    "                                comp = [(aa, rel_aa_comp(AAs, aa)) for aa in subst] # calculate the relative amino acid composition of the possible substitutions\n",
    "                                lowest_comp = min(comp, key=lambda pair: pair[1]) # find the amino acid with the lowest relative composition\n",
    "                                mut_aa = aminoacid + '-' + key + '-' + lowest_comp[0]\n",
    "                                Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                if Diff > Diff_cutoff:\n",
    "                                    AA_mut_list = mut_apply(AA_list, [mut_aa])\n",
    "                                    mut_list.append(mut_aa) # append the mutation to the mutation list\n",
    "                                    break\n",
    "\n",
    "            \n",
    "                        \n",
    "            elif first_entry in neg_corr_list: # checks if the amount of aminoacids should be decreased\n",
    "                for key in free_AA_dict:\n",
    "                    aminoacid = free_AA_dict[key]\n",
    "                    if aminoacid in first_entry: #checks if the aminoacid is present in the first_entry\n",
    "                        AA_subst = conserv_substitution[aminoacid] # list of possible substitutions for the current amino acid\n",
    "                        \n",
    "                        #mutation               \n",
    "                        for n in range(len(deviation)): # iterates through all deviations and takes the first deviation (from pos corr) which can be increaed\n",
    "                            entry = deviation[n][0]\n",
    "                            if entry in pos_corr_list: \n",
    "                                subst = [] #creates a list of possible substitutions that increase one of the amino acids in the first_entry\n",
    "                                for k in AA_subst:\n",
    "                                    if k in entry:\n",
    "                                        subst.append(k)\n",
    "                                if len(subst) == 1: # if only one substitution increases one of the aminoacids this substitution will be used\n",
    "                                    mut_aa = aminoacid + '-' + key + '-' + subst[0]\n",
    "                                    Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                    if Diff > Diff_cutoff:\n",
    "                                        AA_mut_list = mut_apply(AA_list, [mut_aa])\n",
    "                                        mut_list.append(mut_aa) \n",
    "                                        break\n",
    "                                    \n",
    "                                elif len(subst) == 2: # if the aminoacid can be substituted to both aminoacids in first_entry choose the one that has the least frequency\n",
    "                                    comp = [(aa, rel_aa_comp(AAs, aa)) for aa in subst] # calculate the relative amino acid composition of the possible substitutions\n",
    "                                    lowest_comp = min(comp, key=lambda pair: pair[1]) # find the amino acid with the lowest relative composition\n",
    "                                    mut_aa = aminoacid + '-' + key + '-' + lowest_comp[0]\n",
    "                                    Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                    if Diff > Diff_cutoff:\n",
    "                                        AA_mut_list = mut_apply(AA_list, [mut_aa])\n",
    "                                        mut_list.append(mut_aa) # append the mutation to the mutation list\n",
    "                                        break\n",
    "        \n",
    "        \n",
    "        elif 'motif' in first_entry:\n",
    "            if first_entry in pos_corr_list: # checks if the amount of aminoacids should be increased\n",
    "                for key in free_AA_dict:\n",
    "                        aa_back = AAs[key-1]    # aminoacid before the current aminoacid\n",
    "                        aa_current = AAs[key]   # current aminoacid\n",
    "                        aa_for = AAs[key+1]     # aminoacid after the current aminoacid\n",
    "                        \n",
    "                        if (aa_current in first_entry[0]) and (aa_for not in first_entry[1]):    #checks if current aminoacid is part of the motif, and next aminoacid not\n",
    "                            for_subst = conserv_substitution[aa_back] # possible substitutions for the aminoacid before the motif\n",
    "                            for s in for_subst: #iterates through the possible substitutions\n",
    "                                if s in first_entry[1]:\n",
    "                                    mut_aa = f'{aa_back}-{key+1}-{s}'\n",
    "                                    Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                    if Diff > Diff_cutoff:\n",
    "                                        AA_mut_list = mut_apply(AA_list, [mut_aa])\n",
    "                                        mut_list.append(mut_aa)\n",
    "                                        break  \n",
    "                                            \n",
    "                        if (aa_current in first_entry[1]) and (aa_back not in first_entry[0]):   #checks if current aminoacid is part of the motif, and previous aminoacid not\n",
    "                            back_subst = conserv_substitution[aa_for] # possible substitutions for the aminoacid after the motif\n",
    "                            for b in back_subst: #iterates through the possible substitutions\n",
    "                                if b in first_entry[0]:\n",
    "                                    mut_aa = f'{aa_for}-{key-1}-{b}'\n",
    "                                    Diff = mut_live_test(AA_mut_list, [mut_aa], pos_corr, neg_corr, ideal_pos_value, ideal_neg_value) # live test if mutation is benefitical\n",
    "                                    if Diff > Diff_cutoff:\n",
    "                                        AA_mut_list = mut_apply(AA_list, [mut_aa])\n",
    "                                        mut_list.append(mut_aa)\n",
    "                                        break\n",
    "        \n",
    "    return AA_mut_list, mut_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get list of pos/neg correlating features for Meltpoint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\marik\\AppData\\Local\\Temp\\ipykernel_8992\\768696474.py:3: DtypeWarning: Columns (8,21,22,23,25,26,27,28) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  pro_df: pd.DataFrame = pd.read_csv(os.path.join('./data', 'prokaryotes_323columns.csv'))\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "pro_df: pd.DataFrame = pd.read_csv(os.path.join('./data', 'prokaryotes_323columns.csv'))\n",
    "pro_df = pro_df.drop(columns = ['Unnamed: 0', 'Unnamed: 0.1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'AR': 0.19577028919092312, 'VR': 0.16570774850840858, 'LR': 0.22057443760531548, 'LP': 0.19913083779456836, 'MR': 0.10189569954318338, 'FR': 0.1214807452476434, 'WR': 0.09732041140609597, 'NR': 0.10350733491352844, 'YR': 0.11505796446768121, 'YP': 0.09361436465693411, 'ER': 0.1750096588731577, 'EP': 0.15356605906241058, 'RH': 0.10505828276892132, 'RC': 0.0895071066024632, 'RP': 0.14867680267495723, 'RG': 0.1743851008086472, 'PG': 0.15294150099790005, 'R': 0.08506020124285214, 'P': 0.06361660143210504}\n",
      "{'IQ': 0.05490211183163266, 'MQ': 0.03973022298244349, 'MT': 0.054568274035922376, 'FQ': 0.05931526868690351, 'WQ': 0.03515493484535609, 'NQ': 0.041341858352788544, 'QS': 0.057844351285310555, 'QT': 0.06062750041770345, 'QD': 0.06216042713743332, 'QH': 0.04289280620818142, 'QK': 0.06281534035966867, 'QC': 0.027341630041723304, 'ST': 0.07268240233878943, 'SH': 0.054947708129267414, 'SC': 0.03939653196280929, 'TD': 0.0769984781909122, 'TH': 0.0577308572616603, 'TC': 0.04217968109520218, 'Q': 0.022894724682112268, 'T': 0.03773277573559115}\n"
     ]
    }
   ],
   "source": [
    "top_df = pro_df[pro_df['meltPoint'] >= pro_df['meltPoint'].quantile(0.9)]\n",
    "AR_mean = np.mean(top_df['AR'])\n",
    "Ideal_pos_value = {}\n",
    "Ideal_neg_value = {}\n",
    "\n",
    "for key in pos_corr.keys():\n",
    "    Ideal_pos_value[key] = np.mean(top_df[key])\n",
    "\n",
    "for key in neg_corr.keys():\n",
    "    Ideal_neg_value[key] = np.mean(top_df[key])\n",
    "    \n",
    "print(Ideal_pos_value)\n",
    "print(Ideal_neg_value)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "pro_corr = pro_df.corr(numeric_only=True)['meltPoint']\n",
    "pos_corr = pro_corr[pro_corr > 0.32]\n",
    "neg_corr = pro_corr[pro_corr < -0.32]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['IQ', 'MQ', 'MT', 'FQ', 'WQ', 'NQ', 'QS', 'QT', 'QD', 'QH', 'QK', 'QC', 'ST', 'SH', 'SC', 'TD', 'TH', 'TC', 'Qhelix', 'Q', 'T', 'PolarAA']\n"
     ]
    }
   ],
   "source": [
    "pos_corr = list(pro_corr.index)\n",
    "neg_corr = list(neg_corr.index)\n",
    "print(neg_corr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Needs to be mutated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ        -0.429683\n",
      "MQ        -0.507329\n",
      "MT        -0.359458\n",
      "FQ        -0.423630\n",
      "WQ        -0.420057\n",
      "NQ        -0.435562\n",
      "QS        -0.502697\n",
      "QT        -0.529350\n",
      "QD        -0.466556\n",
      "QH        -0.455041\n",
      "QK        -0.418720\n",
      "QC        -0.493738\n",
      "ST        -0.396185\n",
      "SH        -0.327684\n",
      "SC        -0.321686\n",
      "TD        -0.344690\n",
      "TH        -0.346921\n",
      "TC        -0.364763\n",
      "Qhelix    -0.378950\n",
      "Q         -0.469765\n",
      "T         -0.321208\n",
      "PolarAA   -0.406035\n",
      "Name: meltPoint, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "print(pro_corr[pro_corr < -0.32])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
