{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prot_mut(pdb_path, pdb_file, pqr_output_path, locked_aa_pos=None, Deep_mut=True, iterations=100, cutoff_value = -0.005, threshhold = 10000, seed = 0, remove_files = None):\n",
    "    \"\"\"\n",
    "    This function performs protein mutation analysis to improve the thermal stability of a protein, with minimal changes to the structure\n",
    "    (as measured by melting point). It takes a PDB file path, filename, and path for PQR output as input.\n",
    "\n",
    "    Args:\n",
    "        pdb_path (str): Path where the pdb file is stored.\n",
    "        pdb_file (str): Name of the PDB file.\n",
    "        pqr_output_path (str): Path where pqr file will be saved (also fasta and fas).\n",
    "        locked_aa_pos (list, optional): List of amino acid positions that should not be mutated (defaults to None).\n",
    "        Deep_mut (bool, optional): Flag whether to include random mutation in screening\n",
    "            (random + rational, defaults to True).\n",
    "        iterations (int, optional): Number of iterations for rational improvement (defaults to 100).\n",
    "        cutoff_value (float, optional): Cutoff value for mutation selection in rational improvement \n",
    "            (defaults to -0.005).\n",
    "        threshhold (int, optional): Threshold for random mutation acceptance (higher value leads to more mutations, \n",
    "            defaults to 10000).\n",
    "        seed (int, optional): Seed for the random number generator (ensures reproducibility, defaults to 0).\n",
    "\n",
    "    Returns:\n",
    "        list: A list containing three elements:\n",
    "            - Tuple: (WT_SPARC object, best_SPARC object) - Wild-type and best mutated protein SPARC predictions.\n",
    "            - Tuple: (WT amino acid list, best mutated protein amino acid list) - Amino acid sequences.\n",
    "            - Tuple: (WT deviation sum, best mutated protein deviation sum) - Deviations of the protein structures.\n",
    "    \"\"\"\n",
    "    #import functions\n",
    "\n",
    "    from ThERMOS import AA2s4pred\n",
    "    from ThERMOS import diff_weighted\n",
    "    from ThERMOS import mutator_rand\n",
    "    from ThERMOS import mutator_rational\n",
    "    from ThERMOS import functional_aa\n",
    "    from ThERMOS import free_aa\n",
    "    from ThERMOS import Subst_reducer\n",
    "    from ThERMOS import pdb2AA\n",
    "    from ThERMOS import ArraySlice\n",
    "    from ThERMOS import pos_corr, neg_corr, ideal_pos_value, ideal_neg_value, conserv_subst, non_conservative_substitutions\n",
    "\n",
    "    from SPARC import SPARC\n",
    "    \n",
    "    from heapq import heappop, heappush\n",
    "    import heapq    \n",
    "    import os\n",
    "   \n",
    "\n",
    "    #extract protein features\n",
    "    aa_list = pdb2AA(pdb_path, pdb_file)\n",
    "    aa_locked = functional_aa(pdb_path, pdb_file, pqr_output_path)\n",
    "    aa_free = free_aa(pdb_path, pdb_file, aa_locked)\n",
    "    aa_str = ''.join(aa_list)\n",
    "    free_AA_dict = {a: b for a, b in zip(aa_free[:,2], aa_free[:,1] )} # create dictionary from array the key is the absolute aminoacid position and value is the aminoacid\n",
    "    \n",
    "    if locked_aa_pos is not None:\n",
    "        for pos in locked_aa_pos:\n",
    "            free_AA_dict.pop(pos)\n",
    "    \n",
    "    sec_prediction = AA2s4pred('./data/s4pred', pqr_output_path, aa_str, pdb_file, remove_file = remove_files)\n",
    "    possible_substitutions = Subst_reducer(sec_prediction, conserv_subst, free_AA_dict, seed = seed)\n",
    "\n",
    "\n",
    "    #calculate WT deviations\n",
    "    WT_dev_sum, WT_dev = diff_weighted(pos_corr, neg_corr, aa_list, ideal_pos_value, ideal_neg_value, sec_prediction)\n",
    "    \"\"\"\n",
    "     if Deep_mut:\n",
    "        #randoly mutate protein (within given constraints), get top 10 mutations\n",
    "        top_variations = []\n",
    "        largest_variations = []\n",
    "        heappush(top_variations, (WT_dev_sum, WT_dev, aa_str))  # Placeholder for lowest score\n",
    "        heappush(largest_variations, (0, WT_dev, aa_str))  # Placeholder for largest variation\n",
    "        \n",
    "        Mut_seq_str = mutator_rand(aa_list, possible_substitutions, threshhold = threshhold, seed = seed)\n",
    "        for Mut_prot in Mut_seq_str:\n",
    "            Mut_dev_sum, Mut_dev = diff_weighted(pos_corr, neg_corr, Mut_prot, ideal_pos_value, ideal_neg_value, sec_prediction, sort = True)\n",
    "            Str_dev = sum(c1 != c2 for c1, c2 in zip(''.join(Mut_prot), aa_str))\n",
    "\n",
    "            #save the top 10 scores\n",
    "            if len(top_variations) < 11:\n",
    "                heappush(top_variations, (Mut_dev_sum, Mut_dev, Mut_prot)) \n",
    "                    \n",
    "            elif Mut_dev_sum < top_variations[0][0]:\n",
    "                heappush(top_variations, (Mut_dev_sum, Mut_dev, Mut_prot))\n",
    "                heappop(top_variations)\n",
    "            \n",
    "            #saves top 10 largest variations\n",
    "            if len(largest_variations) <11:\n",
    "                heappush(largest_variations, (Str_dev, Mut_dev, Mut_prot))\n",
    "            elif Str_dev > largest_variations[0][0]:\n",
    "                heappush(largest_variations, (Str_dev, Mut_dev, Mut_prot))\n",
    "                heappop(largest_variations)\n",
    "                \n",
    "        heappush(top_variations, (WT_dev_sum, WT_dev, aa_str))\n",
    "        Random_creation = list(heapq.merge(top_variations, largest_variations))\n",
    "        print('Random mutation finished')\n",
    "        \n",
    "        \n",
    "    #define variables for iteration\n",
    "    prev_Mut_prot_list = aa_list    \n",
    "    prev_Mut_dev = WT_dev           \n",
    "    aa_available = aa_free          \n",
    "    \n",
    "    #define variables for best protein#\n",
    "    if Deep_mut:\n",
    "        best_Mut_prot_list = list(heappop(top_variations)[2])\n",
    "        best_Mut_dev_sum = heappop(top_variations)[0]\n",
    "        best_Mut_dev = heappop(top_variations)[1]\n",
    "        best_aa_available = aa_available\n",
    "    else:\n",
    "        best_Mut_prot_list = list(aa_list)\n",
    "        best_Mut_dev_sum = WT_dev_sum\n",
    "        best_Mut_dev = WT_dev\n",
    "        best_aa_available = aa_available\n",
    "        Random_creation = [(WT_dev_sum, WT_dev, aa_str)]\n",
    "    \n",
    "    # Initiate top 5 best variations of rational improvement to calculate melt point\n",
    "    top_top_variations = []\n",
    "    heappush(top_top_variations, (WT_dev_sum, WT_dev, aa_str))\n",
    "    \n",
    "    \n",
    "    #use top 10 mutated sequences and use rational improvement      \n",
    "    best_iteration = 0 #(used to track how many iterations are necessary, currently ~2-3 seems best)\n",
    "    for mut_seq in Random_creation:\n",
    "        prev_Mut_prot_list = list(mut_seq[2])\n",
    "        prev_Mut_dev_sum = mut_seq[0]\n",
    "        prev_Mut_dev = mut_seq[1]\n",
    "\n",
    "        for k in range(iterations):\n",
    "            Mut_prot_list, possible_mutations = mutator_rational(\n",
    "                                                    AA_list = prev_Mut_prot_list, \n",
    "                                                    free_AA = aa_available, \n",
    "                                                    deviation = prev_Mut_dev,\n",
    "                                                    pos_corr = pos_corr, \n",
    "                                                    neg_corr =  neg_corr, \n",
    "                                                    conserv_substitution = possible_substitutions,\n",
    "                                                    ideal_pos_value = ideal_pos_value, \n",
    "                                                    ideal_neg_value = ideal_neg_value,\n",
    "                                                    cutoff = cutoff_value,\n",
    "                                                    sec_prediction = sec_prediction\n",
    "                                                    ) #f_value = cutoff, calculates list of possible mutations\n",
    "            \n",
    "            Mut_dev_sum, Mut_dev = diff_weighted(pos_corr, neg_corr, Mut_prot_list, ideal_pos_value, ideal_neg_value, sec_prediction) # calculate deviation of mutated protein sequence\n",
    "            \n",
    "            if len(top_top_variations) < 6:\n",
    "                heappush(top_top_variations, (Mut_dev_sum, Mut_dev, Mut_prot_list))\n",
    "            elif Mut_dev_sum < top_top_variations[0][0]:\n",
    "                heappush(top_top_variations, (Mut_dev_sum, Mut_dev, Mut_prot_list))\n",
    "                heappop(top_top_variations)\n",
    "                \n",
    "            if abs(best_Mut_dev_sum) > abs(Mut_dev_sum):\n",
    "                best_Mut_prot_list = Mut_prot_list\n",
    "                best_Mut_dev_sum = Mut_dev_sum\n",
    "                best_Mut_dev = Mut_dev  \n",
    "                best_possible_mutations = possible_mutations #get list of best mutations (AA-POS-AA), depreciated, bcs random mutator doesn't output this\n",
    "                best_iteration = str(k+1)\n",
    "                #aa_available = ArraySlice(aa_available, possible_mutations) #updates available aminoacids, so that each aminoacid can only be mutated once\n",
    "                \n",
    "            elif Mut_dev[0][0] == prev_Mut_dev[0][0] and abs(Mut_dev[0][1]-prev_Mut_dev[0][1]) < 0.001:\n",
    "                break\n",
    "                \n",
    "            #update variables for next iteration            \n",
    "            prev_possible_mutations = possible_mutations    #list of mutations (AA-POS-AA) (prev_possible_mutations can be printed if needed)\n",
    "            prev_Mut_prot_list = Mut_prot_list                        #Mutated protein as a list with one AA per entry\n",
    "            prev_Mut_dev = Mut_dev\n",
    "            prev_Mut_dev_sum = Mut_dev_sum\n",
    "    \n",
    "        \n",
    "    #for top_hit in top_top_variations:\n",
    "    #-------SPARC implementation missing--------#\n",
    "    wt_sparc = SPARC(aa_str, pdb_file.split('-')[1], './data', './data/s4pred')\n",
    "    best_temp = wt_sparc[0]\n",
    "    wt_temp = wt_sparc[0]  \n",
    "    best_SPARC = wt_sparc\n",
    "    \n",
    "    #select best mutation based on melt point\n",
    "    for top in top_top_variations:\n",
    "        top_SPARC = SPARC(''.join(top[2]), pdb_file.split('-')[1], './data', './data/s4pred')\n",
    "        top_temp = top_SPARC[0]\n",
    "        if top_SPARC[0] > best_temp:\n",
    "            best_temp = top_SPARC[0]\n",
    "            best_SPARC = top_SPARC\n",
    "            best_Mut_prot_list = top[2]\n",
    "            best_Mut_dev_sum = top[0]\n",
    "            best_Mut_dev = top[1] \"\"\"\n",
    "    \n",
    "    \n",
    "    \n",
    "    #Improvement = WT_dev_sum - best_Mut_dev_sum\n",
    "    \n",
    "    return aa_list, possible_substitutions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pqr file already exists\n",
      "fasta file already exists\n",
      "fas file already exists\n"
     ]
    }
   ],
   "source": [
    "aa_list, poss_subst = prot_mut('./data/pdbs', 'AF-C0H3V2-F1.pdb', './data/test' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def prot_mut_worker(args):\n",
    "    from ThERMOS import prot_mut\n",
    "    pdb_file_path, pqr_output_path, locked_aa_pos, Deep_mut, iterations, cutoff_value, threshhold, seed, remove_files = args\n",
    "    return prot_mut(pdb_file_path, pqr_output_path, locked_aa_pos, Deep_mut, iterations, cutoff_value, threshhold, seed, remove_files)\n",
    "\n",
    "def run_prot_mut_for_multiple_proteins_multiprocessing(pdb_file_paths, pqr_output_path, locked_aa_pos=None, Deep_mut=True, iterations=100, cutoff_value = -0.005, threshhold = 10000, seed = 0, remove_files = None, num_processes=4):\n",
    "    from multiprocessing import Pool\n",
    "    with Pool(processes=num_processes) as pool:\n",
    "        results = pool.starmap(prot_mut_worker, [(pdb_file_path, pqr_output_path, locked_aa_pos, Deep_mut, iterations, cutoff_value, threshhold, seed, remove_files) for pdb_file_path in pdb_file_paths])\n",
    "    return results\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mutational screening of functional proteins\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Mutational screen of all esssential proteins**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pqr file already exists\n",
      "fasta file already exists\n",
      "fas file already exists\n",
      "Random mutation finished\n"
     ]
    }
   ],
   "source": [
    "from ThERMOS import prot_mut\n",
    "test = prot_mut('./data/pdbs', 'AF-C0H3V2-F1.pdb', './data/test', iterations = 1, threshhold = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55.84154306316571\n",
      "55.84154306316571\n",
      "['M', 'Q', 'V', 'L', 'A', 'K', 'E', 'N', 'I', 'K', 'L', 'N', 'Q', 'T', 'V', 'S', 'S', 'K', 'E', 'E', 'A', 'I', 'K', 'L', 'A', 'G', 'Q', 'T', 'L', 'I', 'D', 'N', 'G', 'Y', 'V', 'T', 'E', 'D', 'Y', 'I', 'S', 'K', 'M', 'F', 'E', 'R', 'E', 'E', 'T', 'S', 'S', 'T', 'F', 'M', 'G', 'N', 'F', 'I', 'A', 'I', 'P', 'H', 'G', 'T', 'E', 'E', 'A', 'K', 'S', 'E', 'V', 'L', 'H', 'S', 'G', 'I', 'S', 'I', 'I', 'Q', 'I', 'P', 'E', 'G', 'V', 'E', 'Y', 'G', 'E', 'G', 'N', 'T', 'A', 'K', 'V', 'V', 'F', 'G', 'I', 'A', 'G', 'K', 'N', 'N', 'E', 'H', 'L', 'D', 'I', 'L', 'S', 'N', 'I', 'A', 'I', 'I', 'C', 'S', 'E', 'E', 'E', 'N', 'I', 'E', 'R', 'L', 'I', 'S', 'A', 'K', 'S', 'E', 'E', 'D', 'L', 'I', 'A', 'I', 'F', 'N', 'E', 'V', 'N']\n",
      "['M', 'Q', 'V', 'L', 'A', 'K', 'E', 'N', 'I', 'K', 'L', 'N', 'Q', 'T', 'V', 'S', 'S', 'K', 'E', 'E', 'A', 'I', 'K', 'L', 'A', 'G', 'Q', 'T', 'L', 'I', 'D', 'N', 'G', 'Y', 'V', 'T', 'E', 'D', 'Y', 'I', 'S', 'K', 'M', 'F', 'E', 'R', 'E', 'E', 'T', 'S', 'S', 'T', 'F', 'M', 'G', 'N', 'F', 'I', 'A', 'I', 'P', 'H', 'G', 'T', 'E', 'E', 'A', 'K', 'S', 'E', 'V', 'L', 'H', 'S', 'G', 'I', 'S', 'I', 'I', 'Q', 'I', 'P', 'E', 'G', 'V', 'E', 'Y', 'G', 'E', 'G', 'N', 'T', 'A', 'K', 'V', 'V', 'F', 'G', 'I', 'A', 'G', 'K', 'N', 'N', 'E', 'H', 'L', 'D', 'I', 'L', 'S', 'N', 'I', 'A', 'I', 'I', 'C', 'S', 'E', 'E', 'E', 'N', 'I', 'E', 'R', 'L', 'I', 'S', 'A', 'K', 'S', 'E', 'E', 'D', 'L', 'I', 'A', 'I', 'F', 'N', 'E', 'V', 'N']\n"
     ]
    }
   ],
   "source": [
    "print(test[0][0][0][0])\n",
    "print(test[0][1][0][0])\n",
    "print(test[1][0])\n",
    "print(test[1][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pqr file already exists\n",
      "fasta file already exists\n",
      "fas file already exists\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[1], line 22\u001b[0m\n\u001b[0;32m     20\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m entry \u001b[38;5;129;01min\u001b[39;00m ess_prot_AFdb[\u001b[38;5;241m5\u001b[39m:]:\n\u001b[0;32m     21\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mexists(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(path_pdb, entry)):\n\u001b[1;32m---> 22\u001b[0m         Mutator_dict[entry] \u001b[38;5;241m=\u001b[39m \u001b[43mprot_mut\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpath_pdb\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mentry\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpath_pqr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mDeep_mut\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43miterations\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m100\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcutoff_value\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[38;5;241;43m0.005\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mthreshhold\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m10000\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mseed\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m     23\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m     24\u001b[0m         no_pdb\u001b[38;5;241m.\u001b[39mappend(entry)\n",
      "File \u001b[1;32mc:\\Users\\marik\\OneDrive - bwedu\\Uni HD\\FS 4\\Bioinfo\\topic04_02\\function_mut.py:1319\u001b[0m, in \u001b[0;36mprot_mut\u001b[1;34m(pdb_path, pdb_file, pqr_output_path, locked_aa_pos, Deep_mut, iterations, cutoff_value, threshhold, seed, remove_files)\u001b[0m\n\u001b[0;32m   1316\u001b[0m heappush(largest_variations, (\u001b[38;5;241m0\u001b[39m, WT_dev, aa_str))  \u001b[38;5;66;03m# Placeholder for largest variation\u001b[39;00m\n\u001b[0;32m   1318\u001b[0m Mut_seq_str \u001b[38;5;241m=\u001b[39m mutator_rand(aa_list, possible_substitutions, threshhold \u001b[38;5;241m=\u001b[39m threshhold, seed \u001b[38;5;241m=\u001b[39m seed)\n\u001b[1;32m-> 1319\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m Mut_prot \u001b[38;5;129;01min\u001b[39;00m Mut_seq_str:\n\u001b[0;32m   1320\u001b[0m     Mut_dev_sum, Mut_dev \u001b[38;5;241m=\u001b[39m diff_weighted(pos_corr, neg_corr, Mut_prot, ideal_pos_value, ideal_neg_value, sec_prediction, sort \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m   1321\u001b[0m     Str_dev \u001b[38;5;241m=\u001b[39m \u001b[38;5;28msum\u001b[39m(c1 \u001b[38;5;241m!=\u001b[39m c2 \u001b[38;5;28;01mfor\u001b[39;00m c1, c2 \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(Mut_prot), aa_str))\n",
      "File \u001b[1;32mc:\\Users\\marik\\OneDrive - bwedu\\Uni HD\\FS 4\\Bioinfo\\topic04_02\\function_mut.py:410\u001b[0m, in \u001b[0;36mdiff_weighted\u001b[1;34m(feature_pos, feature_neg, aa, ideal_pos, ideal_neg, sec_prediction, sort, sum_only)\u001b[0m\n\u001b[0;32m    408\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m aa_pos \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(aa) \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m):\n\u001b[0;32m    409\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m helix \u001b[38;5;129;01min\u001b[39;00m helices:\n\u001b[1;32m--> 410\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m aa_pos\u001b[38;5;241m+\u001b[39m\u001b[38;5;241m1\u001b[39m \u001b[38;5;129;01min\u001b[39;00m helix:\n\u001b[0;32m    411\u001b[0m             aa_helix\u001b[38;5;241m.\u001b[39mappend(aa[aa_pos])\n\u001b[0;32m    412\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(aa_helix) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import pickle\n",
    "from ThERMOS import prot_mut\n",
    "data_path = './data/essential_proteins_final.csv'\n",
    "ess_prot_AFdb = (\n",
    "    pd.read_csv(data_path)['AlphaFoldDB']\n",
    "    .dropna()  \n",
    "    .str.replace(';', '')  \n",
    "    .apply(lambda x: f'AF-{x}-F1.pdb') \n",
    ")\n",
    "ess_prot_AFdb = ess_prot_AFdb.tolist()\n",
    "path_pdb = './data/pdbs'\n",
    "path_pqr=  './data/pqrs'\n",
    "output_path = './data/test'\n",
    "\n",
    "Mutator_dict = {}\n",
    "count = 6\n",
    "no_pdb = []\n",
    "for entry in ess_prot_AFdb[5:]:\n",
    "    if os.path.exists(os.path.join(path_pdb, entry)):\n",
    "        Mutator_dict[entry] = prot_mut(path_pdb, entry, path_pqr, Deep_mut=True, iterations=100, cutoff_value = -0.005, threshhold = 10000, seed = 0)\n",
    "    else:\n",
    "        no_pdb.append(entry)\n",
    "        print(f'{entry} not found')\n",
    "    with(open('./data/Mutator_rational_final.pkl', 'wb')) as file:\n",
    "        pickle.dump(Mutator_dict, file)\n",
    "    print(f'{entry} finished ({count}/{len(ess_prot_AFdb)})')\n",
    "    count += 1\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[6, 7, 8]\n"
     ]
    }
   ],
   "source": [
    "test = [1,2,3,4,5,6,7,8]\n",
    "print(test[5:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[86.22309223]\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "\n",
    "with(open('./data/Mutator_rational_02.pkl', 'rb')) as file:\n",
    "    test = pickle.load(file)\n",
    "\n",
    "print(test['AF-O07615-F1.pdb'][0][1][0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Analyze mutation sreen of essential proteins**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "with(open('./data/Essential_genes_mutated.pkl', 'rb')) as f:\n",
    "    Mutator_dict_load = pickle.load(f)\n",
    "with(open('./data/Mutator_dict_13first_bs_error.pkl', 'rb')) as f:\n",
    "    Mutator_dict_13 = pickle.load(f)\n",
    "Mutated_proteins = {**Mutator_dict_load, **Mutator_dict_13}\n",
    "\n",
    "\"\"\" with(open('./data/Mutator_rational_test_top79.pkl', 'rb')) as file:\n",
    "    Mutated_proteins = pickle.load(file) \"\"\"\n",
    "\n",
    "\n",
    "Increase = 0\n",
    "No_improvement = []\n",
    "Mutated_improvement = {}\n",
    "for key in Mutated_proteins:\n",
    "    if Mutated_proteins[key][0][0][0] == Mutated_proteins[key][0][1][0]:\n",
    "        No_improvement.append(key)\n",
    "    else:\n",
    "        Mutated_improvement[key] = Mutated_proteins[key]\n",
    "\n",
    "print(Mutated_improvement.keys())\n",
    "\n",
    "Diff = 0\n",
    "Changed_aa_amount = 0\n",
    "highest_increase = ('name',0)\n",
    "biggest_relative_change = ('name',0)\n",
    "biggest_absolute_chagne = ('name',0)\n",
    "rel_mut_increase = ('name',0)\n",
    "for key in Mutated_improvement:\n",
    "    #Melt point increase\n",
    "    Diff += float(Mutated_improvement[key][0][1][0][0]) - float(Mutated_improvement[key][0][0][0][0])\n",
    "    \n",
    "    #highest increase\n",
    "    if float(Mutated_improvement[key][0][1][0][0]) - float(Mutated_improvement[key][0][0][0][0]) > highest_increase[1]:\n",
    "        highest_increase = (key, float(Mutated_improvement[key][0][1][0][0]) - float(Mutated_improvement[key][0][0][0][0]))\n",
    "    \n",
    "    #percentage of changed aminoacids\n",
    "    changed_aa = 0\n",
    "    for n in range(len(Mutated_improvement[key][1][0])): \n",
    "        if Mutated_improvement[key][1][0][n] != Mutated_improvement[key][1][1][n]:\n",
    "            changed_aa += 1\n",
    "    if changed_aa/len(Mutated_improvement[key][1][0]) > biggest_relative_change[1]:\n",
    "        biggest_relative_change = (key, changed_aa/len(Mutated_improvement[key][1][0]))\n",
    "    if changed_aa > biggest_absolute_chagne[1]:\n",
    "        biggest_absolute_chagne = (key, changed_aa)\n",
    "    Changed_aa_amount += changed_aa/len(Mutated_improvement[key][1][0])\n",
    "    if changed_aa > 0:\n",
    "        if (float(Mutated_improvement[key][0][1][0][0]) - float(Mutated_improvement[key][0][0][0][0]))/float(changed_aa) > rel_mut_increase[1]:\n",
    "            rel_mut_increase = (key, (float(Mutated_improvement[key][0][1][0][0]) - float(Mutated_improvement[key][0][0][0][0]))/float(changed_aa))\n",
    "\n",
    "#Mutation list for edge cases\n",
    "highest_increase_mut_list = []\n",
    "biggest_rel_change_mut_list = []\n",
    "biggest_abs_change_mut_list =[]\n",
    "highest_mut2melt_list = []\n",
    "for key in Mutated_improvement:\n",
    "    if key == highest_increase[0]:\n",
    "        for n in range(len(Mutated_improvement[key][1][0])):\n",
    "            if Mutated_improvement[key][1][0][n] != Mutated_improvement[key][1][1][n]:\n",
    "                highest_increase_mut_list.append(f'{n+1}')\n",
    "    if key == biggest_relative_change[0]:\n",
    "        for n in range(len(Mutated_improvement[key][1][0])):\n",
    "            if Mutated_improvement[key][1][0][n] != Mutated_improvement[key][1][1][n]:\n",
    "                biggest_rel_change_mut_list.append(f'{n+1}')\n",
    "    if key == biggest_absolute_chagne[0]:\n",
    "        for n in range(len(Mutated_improvement[key][1][0])):\n",
    "            if Mutated_improvement[key][1][0][n] != Mutated_improvement[key][1][1][n]:\n",
    "                biggest_abs_change_mut_list.append(f'{n+1}')\n",
    "    if key == rel_mut_increase[0]:\n",
    "        for n in range(len(Mutated_improvement[key][1][0])):\n",
    "            if Mutated_improvement[key][1][0][n] != Mutated_improvement[key][1][1][n]:\n",
    "                highest_mut2melt_list.append(f'{n+1}')\n",
    "\n",
    "\n",
    "#Print results\n",
    "#print(f'Proteins without a pdb file: {len(no_pdb)}')\n",
    "print(f\"Proteins that couldn't be improved: {len(No_improvement)}\")\n",
    "print(f'Proteins that could be improved: {len(Mutated_improvement.keys())} \\n')\n",
    "print(f'Average meltPoint increase: {Diff/len(Mutated_improvement.keys())}')\n",
    "print(f'Average percentage of changed aminoacid per protein: {Changed_aa_amount/len(Mutated_improvement.keys())} \\n')\n",
    "print(f'Highest meltPoint increase: {highest_increase[0]} with {highest_increase[1]} ({Mutated_improvement[highest_increase[0]][0][0][0][0]} -> {Mutated_improvement[highest_increase[0]][0][1][0][0]})\\nWT sequence:      {''.join(Mutated_improvement['AF-P37948-F1.pdb'][1][0])} \\nMutated sequence: {''.join(Mutated_improvement['AF-P37948-F1.pdb'][1][1])}')\n",
    "print(f'{'+'.join(highest_increase_mut_list)}\\n')\n",
    "print(f'Highest percentage of changed aminoacid: {biggest_relative_change[0]} with {biggest_relative_change[1]} with {int(len(Mutated_improvement[biggest_relative_change[0]][1][0])*biggest_relative_change[1])} of {len(Mutated_improvement[biggest_relative_change[0]][1][0])}\\nWT sequence:      {''.join(Mutated_improvement['AF-Q47710-F1.pdb'][1][0])} \\nMutated sequence: {''.join(Mutated_improvement['AF-Q47710-F1.pdb'][1][1])}')\n",
    "print(f'{'+'.join(biggest_rel_change_mut_list)}\\n')\n",
    "print(f'Highest absolute number of changed aminoacid: {biggest_absolute_chagne[0]} with {biggest_absolute_chagne[1]} of {len(Mutated_improvement[biggest_absolute_chagne[0]][1][0])}\\nWT sequence:      {''.join(Mutated_improvement['AF-P46918-F1.pdb'][1][0])} \\nMutated sequence: {''.join(Mutated_improvement['AF-P46918-F1.pdb'][1][1])}')\n",
    "print(f'{'+'.join(biggest_abs_change_mut_list)}\\n')                                                                                                                                                                                                                                         \n",
    "print(f'')\n",
    "print(f'Biggest meltPoint change per mutation {rel_mut_increase[0]} with {rel_mut_increase[1]} ({Mutated_improvement[rel_mut_increase[0]][0][0][0][0]} -> {Mutated_improvement[rel_mut_increase[0]][0][1][0][0]})\\nWT sequence:      {''.join(Mutated_improvement[rel_mut_increase[0]][1][0])} \\nMutated sequence: {''.join(Mutated_improvement[rel_mut_increase[0]][1][1])}')\n",
    "print(f'{'+'.join(highest_mut2melt_list)}\\n')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Create dataframe from mutation results for statistical testing and plotting**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import pandas as pd\n",
    "with(open('./data/Essential_genes_mutated.pkl', 'rb')) as f:\n",
    "    Mutator_dict_load = pickle.load(f)\n",
    "with(open('./data/Mutator_dict_13first_bs_error.pkl', 'rb')) as f:\n",
    "    Mutator_dict_13 = pickle.load(f)\n",
    "Mutated_proteins = {**Mutator_dict_load, **Mutator_dict_13}\n",
    "Mutated_df = pd.DataFrame(columns = ['Protein', 'WT_prediction', 'Mut_prediction', 'WT_sequence', 'Mut_sequence', 'Mutations'])\n",
    "for key in Mutated_proteins:\n",
    "    wt_prediction = Mutated_proteins[key][0][0][0][0]\n",
    "    mut_prediction = Mutated_proteins[key][0][1][0][0]\n",
    "    wt_seq = ''.join(Mutated_proteins[key][1][0])\n",
    "    mut_seq = ''.join(Mutated_proteins[key][1][1])\n",
    "    mutations_list = []\n",
    "    for i in range(len(wt_seq)):\n",
    "        if wt_seq[i] != mut_seq[i]:\n",
    "            mutations_list.append(f'{wt_seq[i]}{i+1}{mut_seq[i]}')\n",
    "    new_row = [key.split('-')[1], wt_prediction, mut_prediction, wt_seq, mut_seq, mutations_list]\n",
    "    #print(new_row)\n",
    "    Mutated_df.loc[len(Mutated_df)] = new_row\n",
    "\n",
    "print(Mutated_df.loc[1])\n",
    "\n",
    "with(open('./data/Mutated_proteins_df.pkl', 'wb')) as f:\n",
    "    pickle.dump(Mutated_df, f)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
